<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Giovanny Covarrubias-Pazaran" />

<meta name="date" content="2025-11-19" />

<title>Fitting genotype by environment models in lme4breeding</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Fitting genotype by environment models in
lme4breeding</h1>
<h4 class="author">Giovanny Covarrubias-Pazaran</h4>
<h4 class="date">2025-11-19</h4>



<p>The purpose of this vignette is to show how to fit different genotype
by environment (GxE) models using the lme4breeding package:</p>
<ol style="list-style-type: decimal">
<li>Multienvironment model: Main effect model (M) 2.1) Multienvironment
model: Diagonal model (DG) 2.2) Heterogeneous residuals</li>
<li>Multienvironment model: Compund symmetry model (CS) 3.2)
Multienvironment model: Main + Diagonal model (M+DG)</li>
<li>Multienvironment model: Unstructured model (US)</li>
<li>Multienvironment model: Random regression model (RR) 5.2)
Multienvironment model: Finlay-Wilkinson regression</li>
<li>Multienvironment model: Factor analytic (reduced rank) model
(FA)</li>
<li>Two stage analysis</li>
<li>How to deal with massive GxE genomic models or massive number of
individuals</li>
</ol>
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>When the breeder decides to run a trial and apply selection in a
single environment (whether because the amount of seed is a limitation
or there’s no availability for a location) the breeder takes the risk of
selecting material for a target population of environments (TPEs) using
an environment that is not representative of the larger TPE. Therefore,
many breeding programs try to base their selection decision on
multi-environment trial (MET) data. Models could be adjusted by adding
additional information like spatial information, experimental design
information, etc. In this tutorial we will focus mainly on the
covariance structures for GxE and the incorporation of relationship
matrices for the genotype effect.</p>
</div>
<div id="met-main-effect-model" class="section level2">
<h2>1) MET: main effect model</h2>
<p>A multi-environment model is the one that is fitted when the breeding
program can afford more than one location in order to cover the target
population of environments (TPE). The main effect model assumes that GxE
doesn’t exist and that the main genotype effect plus the fixed effect
for environment is enough to predict the genotype effect in all
locations of interest.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">data</span>(DT_example, <span class="at">package=</span><span class="st">&quot;enhancer&quot;</span>)</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a>DT <span class="ot">&lt;-</span> DT_example</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a>A <span class="ot">&lt;-</span> A_example</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a></span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a>ansMain <span class="ot">&lt;-</span> <span class="fu">lmeb</span>(Yield <span class="sc">~</span> Env <span class="sc">+</span> (<span class="dv">1</span><span class="sc">|</span>Name),</span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a>                        <span class="at">relmat =</span> <span class="fu">list</span>(<span class="at">Name =</span> A ),</span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a>                         <span class="at">verbose =</span> <span class="dv">0</span><span class="dt">L</span>, <span class="at">trace=</span><span class="dv">0</span><span class="dt">L</span>, <span class="at">data=</span>DT)</span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a>vc <span class="ot">&lt;-</span> <span class="fu">VarCorr</span>(ansMain); <span class="fu">print</span>(vc,<span class="at">comp=</span><span class="fu">c</span>(<span class="st">&quot;Variance&quot;</span>))</span></code></pre></div>
<pre><code>##  Groups   Name        Variance
##  Name     (Intercept) 4.8559  
##  Residual             8.1086</code></pre>
<p>In this model, the only term to be estimated is the one for the
germplasm (here called <code>Name</code>). For the sake of example we
have added a relationship matrix among the levels of the random effect
<code>Name</code>. This is just a diagonal matrix with as many rows and
columns as levels present in the random effect <code>Name</code>, but
any other non-diagonal relationship matrix could be used.</p>
</div>
<div id="met-diagonal-model-dg" class="section level2">
<h2>2.1) MET: diagonal model (DG)</h2>
<p>A multi-environment model is the one that is fitted when the breeding
program can afford more than one location. The diagonal model assumes
that GxE exists and that the genotype variation is expressed differently
at each location, therefore fitting a variance component for the
genotype effect at each location. The main drawback is that this model
assumes no covariance among locations, as if genotypes were independent
(despite the fact that is the same genotypes). The fixed effect for
environment plus the location-specific BLUP is used to predict the
genotype effect in each locations of interest.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a>ansDG <span class="ot">&lt;-</span> <span class="fu">lmeb</span>(Yield <span class="sc">~</span> Env <span class="sc">+</span> (<span class="dv">0</span><span class="sc">+</span>Env <span class="sc">||</span> Name),</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>                      <span class="at">relmat =</span> <span class="fu">list</span>(<span class="at">Name =</span> A ),</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>                       <span class="at">verbose =</span> <span class="dv">0</span><span class="dt">L</span>, <span class="at">trace=</span><span class="dv">0</span><span class="dt">L</span>, <span class="at">data=</span>DT)</span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>vc <span class="ot">&lt;-</span> <span class="fu">VarCorr</span>(ansDG); <span class="fu">print</span>(vc,<span class="at">comp=</span><span class="fu">c</span>(<span class="st">&quot;Variance&quot;</span>))</span></code></pre></div>
<pre><code>##  Groups   Name    Variance
##  Name     CA.2011 17.4934 
##  Name.1   CA.2012  5.3376 
##  Name.2   CA.2013  7.8839 
##  Residual          4.3806</code></pre>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a>ve <span class="ot">&lt;-</span> <span class="fu">attr</span>(vc, <span class="st">&quot;sc&quot;</span>)<span class="sc">^</span><span class="dv">2</span>; ve</span></code></pre></div>
<pre><code>## [1] 4.380609</code></pre>
</div>
<div id="heterogeneous-residuals" class="section level2">
<h2>2,2) Heterogeneous residuals</h2>
<p>The lme4breeding has a reserved word named ‘unitsR’ to call for
residuals. We will show how to fit heterogeneous residuals at the level
of each environment with the previous diagonal model.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a>ansCSDG <span class="ot">&lt;-</span> <span class="fu">lmeb</span>(Yield <span class="sc">~</span> Env <span class="sc">+</span> (<span class="dv">0</span><span class="sc">+</span>Env <span class="sc">||</span> Name) <span class="sc">+</span> (<span class="dv">0</span><span class="sc">+</span>Env <span class="sc">||</span> unitsR),</span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a>                      <span class="at">relmat =</span> <span class="fu">list</span>(<span class="at">Name =</span> A ),</span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a>                       <span class="at">verbose =</span> <span class="dv">0</span><span class="dt">L</span>, <span class="at">trace=</span><span class="dv">0</span><span class="dt">L</span>, <span class="at">data=</span>DT)</span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a>vc <span class="ot">&lt;-</span> <span class="fu">VarCorr</span>(ansCSDG); <span class="fu">print</span>(vc,<span class="at">comp=</span><span class="fu">c</span>(<span class="st">&quot;Variance&quot;</span>))</span></code></pre></div>
<pre><code>##  Groups   Name    Variance 
##  unitsR   CA.2013  0.041849
##  unitsR.1 CA.2012  3.148196
##  unitsR.2 CA.2011  2.438074
##  Name     CA.2013  8.791326
##  Name.1   CA.2012  4.596127
##  Name.2   CA.2011 17.215621
##  Residual          2.515180</code></pre>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a>ve <span class="ot">&lt;-</span> <span class="fu">attr</span>(vc, <span class="st">&quot;sc&quot;</span>)<span class="sc">^</span><span class="dv">2</span>; ve</span></code></pre></div>
<pre><code>## [1] 2.51518</code></pre>
</div>
<div id="met-compund-symmetry-model-cs" class="section level2">
<h2>3) MET: compund symmetry model (CS)</h2>
<p>A multi-environment model is the one that is fitted when the breeding
program can afford more than one location. The compound symmetry model
assumes that GxE exists and that a main genotype variance-covariance
component is expressed across all location. In addition, it assumes that
a main genotype-by-environment variance is expressed across all
locations. The main drawback is that the model assumes the same variance
and covariance among locations. The fixed effect for environment plus
the main effect for BLUP plus genotype-by-environment effect is used to
predict the genotype effect in each location of interest.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a>DT<span class="sc">$</span>EnvName <span class="ot">&lt;-</span> <span class="fu">paste</span>(DT<span class="sc">$</span>Env, DT<span class="sc">$</span>Name, <span class="at">sep =</span> <span class="st">&quot;:&quot;</span>)</span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a>E <span class="ot">&lt;-</span> Matrix<span class="sc">::</span><span class="fu">Diagonal</span>(<span class="fu">length</span>(<span class="fu">unique</span>(DT<span class="sc">$</span>Env)));</span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a><span class="fu">colnames</span>(E) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(E) <span class="ot">&lt;-</span> <span class="fu">unique</span>(DT<span class="sc">$</span>Env);E</span></code></pre></div>
<pre><code>## 3 x 3 diagonal matrix of class &quot;ddiMatrix&quot;
##         CA.2013 CA.2011 CA.2012
## CA.2013       1       .       .
## CA.2011       .       1       .
## CA.2012       .       .       1</code></pre>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a>EA <span class="ot">&lt;-</span> Matrix<span class="sc">::</span><span class="fu">kronecker</span>(E,A, <span class="at">make.dimnames =</span> <span class="cn">TRUE</span>)</span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a>ansCS <span class="ot">&lt;-</span> <span class="fu">lmeb</span>(Yield <span class="sc">~</span> Env <span class="sc">+</span> (<span class="dv">1</span><span class="sc">|</span>Name) <span class="sc">+</span> (<span class="dv">1</span><span class="sc">|</span>EnvName),</span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a>                    <span class="at">relmat =</span> <span class="fu">list</span>(<span class="at">Name =</span> A, <span class="at">EnvName=</span>  EA ),</span>
<span id="cb13-4"><a href="#cb13-4" tabindex="-1"></a>                     <span class="at">verbose =</span> <span class="dv">0</span><span class="dt">L</span>, <span class="at">trace=</span><span class="dv">0</span><span class="dt">L</span>, <span class="at">data=</span>DT)</span>
<span id="cb13-5"><a href="#cb13-5" tabindex="-1"></a>vc <span class="ot">&lt;-</span> <span class="fu">VarCorr</span>(ansCS); <span class="fu">print</span>(vc,<span class="at">comp=</span><span class="fu">c</span>(<span class="st">&quot;Variance&quot;</span>))</span></code></pre></div>
<pre><code>##  Groups   Name        Variance
##  EnvName  (Intercept) 5.1732  
##  Name     (Intercept) 3.6819  
##  Residual             4.3662</code></pre>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a>ve <span class="ot">&lt;-</span> <span class="fu">attr</span>(vc, <span class="st">&quot;sc&quot;</span>)<span class="sc">^</span><span class="dv">2</span>; ve</span></code></pre></div>
<pre><code>## [1] 4.366211</code></pre>
<div id="met-main-diagonal-mdg" class="section level3">
<h3>3.2) MET: main + diagonal (M+DG)</h3>
<p>A multi-environment model is the one that is fitted when the breeding
program can afford more than one location. The compound symmetry model
assumes that GxE exists and that a main genotype variance-covariance
component is expressed across all location. In addition, it assumes that
a main genotype-by-environment variance is expressed across all
locations. The main drawback is that the model assumes the same variance
and covariance among locations. The fixed effect for environment plus
the main effect for BLUP plus genotype-by-environment effect is used to
predict the genotype effect in each location of interest.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a>ansCSDG <span class="ot">&lt;-</span> <span class="fu">lmeb</span>(Yield <span class="sc">~</span> Env <span class="sc">+</span> (Env <span class="sc">||</span> Name),</span>
<span id="cb17-2"><a href="#cb17-2" tabindex="-1"></a>                      <span class="at">relmat =</span> <span class="fu">list</span>(<span class="at">Name =</span> A ),</span>
<span id="cb17-3"><a href="#cb17-3" tabindex="-1"></a>                       <span class="at">verbose =</span> <span class="dv">0</span><span class="dt">L</span>, <span class="at">trace=</span><span class="dv">0</span><span class="dt">L</span>, <span class="at">data=</span>DT)</span>
<span id="cb17-4"><a href="#cb17-4" tabindex="-1"></a>vc <span class="ot">&lt;-</span> <span class="fu">VarCorr</span>(ansCSDG); <span class="fu">print</span>(vc,<span class="at">comp=</span><span class="fu">c</span>(<span class="st">&quot;Variance&quot;</span>))</span></code></pre></div>
<pre><code>##  Groups   Name        Variance
##  Name     (Intercept)  2.9637 
##  Name.1   CA.2011     10.4264 
##  Name.2   CA.2012      2.6589 
##  Name.3   CA.2013      5.7017 
##  Residual              4.3976</code></pre>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a>ve <span class="ot">&lt;-</span> <span class="fu">attr</span>(vc, <span class="st">&quot;sc&quot;</span>)<span class="sc">^</span><span class="dv">2</span>; ve</span></code></pre></div>
<pre><code>## [1] 4.397594</code></pre>
</div>
</div>
<div id="met-unstructured-model-us" class="section level2">
<h2>4) MET: unstructured model (US)</h2>
<p>A multi-environment model is the one that is fitted when the breeding
program can afford more than one location. The unstructured model is the
most flexible model assuming that GxE exists and that an
environment-specific variance exists in addition to as many covariances
for each environment-to-environment combinations. The main drawback is
that is difficult to make this models converge because of the large
number of variance components, the fact that some of these variance or
covariance components are zero, and the difficulty in choosing good
starting values. The fixed effect for environment plus the environment
specific BLUP (adjusted by covariances) is used to predict the genotype
effect in each location of interest.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a>ansUS <span class="ot">&lt;-</span> <span class="fu">lmeb</span>(Yield <span class="sc">~</span> Env <span class="sc">+</span> (<span class="dv">0</span><span class="sc">+</span>Env <span class="sc">|</span> Name),</span>
<span id="cb21-2"><a href="#cb21-2" tabindex="-1"></a>                    <span class="at">relmat =</span> <span class="fu">list</span>(<span class="at">Name =</span> A ),</span>
<span id="cb21-3"><a href="#cb21-3" tabindex="-1"></a>                     <span class="at">verbose =</span> <span class="dv">0</span><span class="dt">L</span>, <span class="at">trace=</span><span class="dv">0</span><span class="dt">L</span>, <span class="at">data=</span>DT)</span>
<span id="cb21-4"><a href="#cb21-4" tabindex="-1"></a>vc <span class="ot">&lt;-</span> <span class="fu">VarCorr</span>(ansUS); <span class="fu">print</span>(vc,<span class="at">comp=</span><span class="fu">c</span>(<span class="st">&quot;Variance&quot;</span>))</span></code></pre></div>
<pre><code>##  Groups   Name    Variance Cov          
##  Name     CA.2011 15.9930               
##           CA.2012  5.2743   6.172       
##           CA.2013  7.6897   6.366  0.375
##  Residual          4.3862</code></pre>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" tabindex="-1"></a>ve <span class="ot">&lt;-</span> <span class="fu">attr</span>(vc, <span class="st">&quot;sc&quot;</span>)<span class="sc">^</span><span class="dv">2</span>; ve</span></code></pre></div>
<pre><code>## [1] 4.38618</code></pre>
</div>
<div id="met-random-regression-model" class="section level2">
<h2>5) MET: random regression model</h2>
<p>A multi-environment model is the one that is fitted when the breeding
program can afford more than one location. The random regression model
assumes that the environment can be seen as a continuous variable and
therefore a variance component for the intercept and a variance
component for the slope can be fitted. The number of variance components
will depend on the order of the Legendre polynomial fitted.</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" tabindex="-1"></a><span class="co"># library(orthopolynom)</span></span>
<span id="cb25-2"><a href="#cb25-2" tabindex="-1"></a><span class="co"># DT$EnvN &lt;- as.numeric(as.factor(DT$Env))</span></span>
<span id="cb25-3"><a href="#cb25-3" tabindex="-1"></a><span class="co"># # add new columns to dataset</span></span>
<span id="cb25-4"><a href="#cb25-4" tabindex="-1"></a><span class="co"># Z &lt;- with(DT, smm(leg(EnvN,1)) )</span></span>
<span id="cb25-5"><a href="#cb25-5" tabindex="-1"></a><span class="co"># for(i in 1:ncol(Z)){DT[,colnames(Z)[i]] &lt;- Z[,i]}</span></span>
<span id="cb25-6"><a href="#cb25-6" tabindex="-1"></a><span class="co"># ansRR &lt;- lmeb(Yield ~ Env + (0+leg0+leg1| Name),</span></span>
<span id="cb25-7"><a href="#cb25-7" tabindex="-1"></a><span class="co">#                   relmat = list(Name = A ),</span></span>
<span id="cb25-8"><a href="#cb25-8" tabindex="-1"></a><span class="co">#                   verbose = 0L, trace=0L,</span></span>
<span id="cb25-9"><a href="#cb25-9" tabindex="-1"></a><span class="co">#                   data=DT)</span></span>
<span id="cb25-10"><a href="#cb25-10" tabindex="-1"></a><span class="co"># vc &lt;- VarCorr(ansRR); print(vc,comp=c(&quot;Variance&quot;))</span></span>
<span id="cb25-11"><a href="#cb25-11" tabindex="-1"></a><span class="co"># ve &lt;- attr(vc, &quot;sc&quot;)^2; ve</span></span></code></pre></div>
<p>In addition, we can fit this without covariance between random slope
and intercept:</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" tabindex="-1"></a><span class="co"># ansRR2 &lt;- lmeb(Yield ~ Env + (0+leg0+leg1|| Name),</span></span>
<span id="cb26-2"><a href="#cb26-2" tabindex="-1"></a><span class="co">#                   relmat = list(Name = A ),</span></span>
<span id="cb26-3"><a href="#cb26-3" tabindex="-1"></a><span class="co">#                   verbose = 0L, trace=0L,</span></span>
<span id="cb26-4"><a href="#cb26-4" tabindex="-1"></a><span class="co">#                   data=DT)</span></span>
<span id="cb26-5"><a href="#cb26-5" tabindex="-1"></a><span class="co"># vc &lt;- VarCorr(ansRR2); print(vc,comp=c(&quot;Variance&quot;))</span></span>
<span id="cb26-6"><a href="#cb26-6" tabindex="-1"></a><span class="co"># ve &lt;- attr(vc, &quot;sc&quot;)^2; ve</span></span></code></pre></div>
<div id="finlay-wilkinson-regression" class="section level3">
<h3>5.2) Finlay-Wilkinson regression</h3>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" tabindex="-1"></a><span class="fu">data</span>(DT_h2, <span class="at">package=</span><span class="st">&quot;enhancer&quot;</span>)</span>
<span id="cb27-2"><a href="#cb27-2" tabindex="-1"></a>DT <span class="ot">&lt;-</span> DT_h2</span>
<span id="cb27-3"><a href="#cb27-3" tabindex="-1"></a><span class="do">## build the environmental index</span></span>
<span id="cb27-4"><a href="#cb27-4" tabindex="-1"></a>ei <span class="ot">&lt;-</span> <span class="fu">aggregate</span>(y<span class="sc">~</span>Env, <span class="at">data=</span>DT,<span class="at">FUN=</span>mean)</span>
<span id="cb27-5"><a href="#cb27-5" tabindex="-1"></a><span class="fu">colnames</span>(ei)[<span class="dv">2</span>] <span class="ot">&lt;-</span> <span class="st">&quot;envIndex&quot;</span></span>
<span id="cb27-6"><a href="#cb27-6" tabindex="-1"></a>ei<span class="sc">$</span>envIndex <span class="ot">&lt;-</span> ei<span class="sc">$</span>envIndex <span class="sc">-</span> <span class="fu">mean</span>(ei<span class="sc">$</span>envIndex,<span class="at">na.rm=</span><span class="cn">TRUE</span>) <span class="co"># center the envIndex to have clean VCs</span></span>
<span id="cb27-7"><a href="#cb27-7" tabindex="-1"></a>ei <span class="ot">&lt;-</span> ei[<span class="fu">with</span>(ei, <span class="fu">order</span>(envIndex)), ]</span>
<span id="cb27-8"><a href="#cb27-8" tabindex="-1"></a><span class="do">## add the environmental index to the original dataset</span></span>
<span id="cb27-9"><a href="#cb27-9" tabindex="-1"></a>DT2 <span class="ot">&lt;-</span> <span class="fu">merge</span>(DT,ei, <span class="at">by=</span><span class="st">&quot;Env&quot;</span>)</span>
<span id="cb27-10"><a href="#cb27-10" tabindex="-1"></a>DT2 <span class="ot">&lt;-</span> DT2[<span class="fu">with</span>(DT2, <span class="fu">order</span>(Name)), ]</span>
<span id="cb27-11"><a href="#cb27-11" tabindex="-1"></a></span>
<span id="cb27-12"><a href="#cb27-12" tabindex="-1"></a>ansFW <span class="ot">&lt;-</span> <span class="fu">lmeb</span>(y<span class="sc">~</span> Env <span class="sc">+</span> (envIndex <span class="sc">||</span> Name),  <span class="at">verbose =</span> <span class="dv">0</span><span class="dt">L</span>, <span class="at">trace=</span><span class="dv">0</span><span class="dt">L</span>, <span class="at">data=</span>DT2)</span>
<span id="cb27-13"><a href="#cb27-13" tabindex="-1"></a>vc <span class="ot">&lt;-</span> <span class="fu">VarCorr</span>(ansFW); <span class="fu">print</span>(vc,<span class="at">comp=</span><span class="fu">c</span>(<span class="st">&quot;Variance&quot;</span>))</span></code></pre></div>
<pre><code>##  Groups   Name        Variance
##  Name     (Intercept) 2.632643
##  Name.1   envIndex    0.041216
##  Residual             7.046535</code></pre>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" tabindex="-1"></a>ve <span class="ot">&lt;-</span> <span class="fu">attr</span>(vc, <span class="st">&quot;sc&quot;</span>)<span class="sc">^</span><span class="dv">2</span>; ve</span></code></pre></div>
<pre><code>## [1] 7.046535</code></pre>
<p>Alternatively, you can also add the covariance between both the main
effect (intercept) and the sensitivity (slope).</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" tabindex="-1"></a>ansFW2 <span class="ot">&lt;-</span> <span class="fu">lmeb</span>(y<span class="sc">~</span> Env <span class="sc">+</span> (envIndex <span class="sc">|</span> Name),  <span class="at">verbose =</span> <span class="dv">0</span><span class="dt">L</span>, <span class="at">trace=</span><span class="dv">0</span><span class="dt">L</span>, <span class="at">data=</span>DT2)</span></code></pre></div>
<pre><code>## boundary (singular) fit: see help(&#39;isSingular&#39;)</code></pre>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" tabindex="-1"></a>vc <span class="ot">&lt;-</span> <span class="fu">VarCorr</span>(ansFW2); <span class="fu">print</span>(vc,<span class="at">comp=</span><span class="fu">c</span>(<span class="st">&quot;Variance&quot;</span>))</span></code></pre></div>
<pre><code>##  Groups   Name        Variance Cov  
##  Name     (Intercept) 2.75593       
##           envIndex    0.04271  0.343
##  Residual             7.03825</code></pre>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" tabindex="-1"></a>ve <span class="ot">&lt;-</span> <span class="fu">attr</span>(vc, <span class="st">&quot;sc&quot;</span>)<span class="sc">^</span><span class="dv">2</span>; ve</span></code></pre></div>
<pre><code>## [1] 7.038247</code></pre>
</div>
</div>
<div id="factor-analytic-reduced-rank-model" class="section level2">
<h2>6) Factor analytic (reduced rank) model</h2>
<p>When the number of environments where genotypes are evaluated is big
and we want to consider the genetic covariance between environments and
location-specific variance components we cannot fit an unstructured
covariance in the model since the number of parameters is too big and
the matrix can become non-full rank leading to singularities. In those
cases is suggested a dimensionality reduction technique. Among those the
factor analytic structures proposed by many research groups (Piepho,
Smith, Cullis, Thompson, Meyer, etc.) are the way to go. lme4breeding
has a reduced-rank factor analytic implementation available through the
rrm() function. Here we show an example of how to fit the model:</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" tabindex="-1"></a><span class="fu">data</span>(DT_h2, <span class="at">package=</span><span class="st">&quot;enhancer&quot;</span>)</span>
<span id="cb37-2"><a href="#cb37-2" tabindex="-1"></a>DT <span class="ot">&lt;-</span> DT_h2</span>
<span id="cb37-3"><a href="#cb37-3" tabindex="-1"></a>DT<span class="ot">=</span>DT[<span class="fu">with</span>(DT, <span class="fu">order</span>(Env)), ]</span>
<span id="cb37-4"><a href="#cb37-4" tabindex="-1"></a><span class="fu">head</span>(DT)</span></code></pre></div>
<pre><code>##          Name     Env Loc Year     Block  y
## 67   MSL007-B CA.2011  CA 2011 CA.2011.2  5
## 105  MSL007-B CA.2011  CA 2011 CA.2011.1  6
## 308  MSK061-4 CA.2011  CA 2011 CA.2011.2  9
## 393  MSK061-4 CA.2011  CA 2011 CA.2011.1 10
## 469 MSR169-8Y CA.2011  CA 2011 CA.2011.1 11
## 471     NY148 CA.2011  CA 2011 CA.2011.1 11</code></pre>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" tabindex="-1"></a>indNames <span class="ot">&lt;-</span> <span class="fu">na.omit</span>(<span class="fu">unique</span>(DT<span class="sc">$</span>Name))</span>
<span id="cb39-2"><a href="#cb39-2" tabindex="-1"></a>A <span class="ot">&lt;-</span> <span class="fu">diag</span>(<span class="fu">length</span>(indNames))</span>
<span id="cb39-3"><a href="#cb39-3" tabindex="-1"></a><span class="fu">rownames</span>(A) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(A) <span class="ot">&lt;-</span> indNames</span>
<span id="cb39-4"><a href="#cb39-4" tabindex="-1"></a></span>
<span id="cb39-5"><a href="#cb39-5" tabindex="-1"></a><span class="co"># fit diagonal model first to produce a two-way table of genotype by env BLUPs</span></span>
<span id="cb39-6"><a href="#cb39-6" tabindex="-1"></a>Z <span class="ot">&lt;-</span> <span class="fu">with</span>(DT, <span class="fu">smm</span>(Env))</span>
<span id="cb39-7"><a href="#cb39-7" tabindex="-1"></a>diagFormula <span class="ot">&lt;-</span> <span class="fu">paste0</span>( <span class="st">&quot;y ~ Env + (0+&quot;</span>, <span class="fu">paste</span>(<span class="fu">colnames</span>(Z), <span class="at">collapse =</span> <span class="st">&quot;+&quot;</span>), <span class="st">&quot;|| Name)&quot;</span>)</span>
<span id="cb39-8"><a href="#cb39-8" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(Z)){DT[,<span class="fu">colnames</span>(Z)[i]] <span class="ot">&lt;-</span> Z[,i]}</span>
<span id="cb39-9"><a href="#cb39-9" tabindex="-1"></a>ans1a <span class="ot">&lt;-</span> <span class="fu">lmeb</span>(<span class="fu">as.formula</span>(diagFormula),</span>
<span id="cb39-10"><a href="#cb39-10" tabindex="-1"></a>                  <span class="at">relmat =</span> <span class="fu">list</span>(<span class="at">Name =</span> A ),</span>
<span id="cb39-11"><a href="#cb39-11" tabindex="-1"></a>                  <span class="at">verbose =</span> <span class="dv">0</span><span class="dt">L</span>, <span class="at">trace=</span><span class="dv">0</span><span class="dt">L</span>, </span>
<span id="cb39-12"><a href="#cb39-12" tabindex="-1"></a>                  <span class="at">data=</span>DT)</span>
<span id="cb39-13"><a href="#cb39-13" tabindex="-1"></a>vc <span class="ot">&lt;-</span> <span class="fu">VarCorr</span>(ans1a); </span>
<span id="cb39-14"><a href="#cb39-14" tabindex="-1"></a>H0 <span class="ot">&lt;-</span> <span class="fu">ranef</span>(ans1a)<span class="sc">$</span>Name <span class="co"># GxE table</span></span></code></pre></div>
<pre><code>## Rotating back conditional variance using Cholesky factors</code></pre>
<pre><code>## Rotating back BLUPs by transpose of Cholesky (u=L&#39;u*) for: Name</code></pre>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" tabindex="-1"></a><span class="co"># Use the GxE table to obtain loadings and build loadings columns in dataset</span></span>
<span id="cb42-2"><a href="#cb42-2" tabindex="-1"></a>nPC<span class="ot">=</span><span class="dv">3</span></span>
<span id="cb42-3"><a href="#cb42-3" tabindex="-1"></a>Z <span class="ot">&lt;-</span> <span class="fu">with</span>(DT,  <span class="fu">rrm</span>(Env, <span class="at">H =</span> H0, <span class="at">nPC =</span> nPC) ) </span>
<span id="cb42-4"><a href="#cb42-4" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(Z)){DT[,<span class="fu">colnames</span>(Z)[i]] <span class="ot">&lt;-</span> Z[,i]}</span>
<span id="cb42-5"><a href="#cb42-5" tabindex="-1"></a><span class="co"># fit the FA model (rr + diag) to calculate factor scores</span></span>
<span id="cb42-6"><a href="#cb42-6" tabindex="-1"></a>ansFA <span class="ot">&lt;-</span> <span class="fu">lmeb</span>(y <span class="sc">~</span> Env <span class="sc">+</span> (<span class="dv">0</span><span class="sc">+</span>PC1<span class="sc">+</span>PC2<span class="sc">+</span>PC3<span class="sc">|</span>Name) <span class="sc">+</span> (<span class="dv">0</span><span class="sc">+</span>Env<span class="sc">||</span> Name),</span>
<span id="cb42-7"><a href="#cb42-7" tabindex="-1"></a>                  <span class="at">relmat =</span> <span class="fu">list</span>(<span class="at">Name =</span> A ),</span>
<span id="cb42-8"><a href="#cb42-8" tabindex="-1"></a>                  <span class="at">verbose =</span> <span class="dv">0</span><span class="dt">L</span>, <span class="at">trace=</span><span class="dv">0</span><span class="dt">L</span>, <span class="at">data=</span>DT)</span>
<span id="cb42-9"><a href="#cb42-9" tabindex="-1"></a></span>
<span id="cb42-10"><a href="#cb42-10" tabindex="-1"></a>u <span class="ot">&lt;-</span> <span class="fu">ranef</span>(ansFA)<span class="sc">$</span>Name <span class="co"># all BLUPs</span></span></code></pre></div>
<pre><code>## Rotating back conditional variance using Cholesky factors
## Rotating back BLUPs by transpose of Cholesky (u=L&#39;u*) for: Name</code></pre>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" tabindex="-1"></a>scores <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(u[,<span class="dv">1</span><span class="sc">:</span>nPC])  <span class="co"># extract factor scores</span></span>
<span id="cb44-2"><a href="#cb44-2" tabindex="-1"></a>loadings<span class="ot">=</span><span class="fu">with</span>(DT, <span class="fu">rrm</span>(Env, <span class="at">nPC =</span> nPC, <span class="at">H =</span> H0, <span class="co"># lodings (latent covars)</span></span>
<span id="cb44-3"><a href="#cb44-3" tabindex="-1"></a>                      <span class="at">returnGamma =</span> <span class="cn">TRUE</span>) )<span class="sc">$</span>Gamma</span>
<span id="cb44-4"><a href="#cb44-4" tabindex="-1"></a></span>
<span id="cb44-5"><a href="#cb44-5" tabindex="-1"></a>vc <span class="ot">&lt;-</span> <span class="fu">VarCorr</span>(ansFA); <span class="fu">print</span>(vc,<span class="at">comp=</span><span class="fu">c</span>(<span class="st">&quot;Variance&quot;</span>)) <span class="co"># extract all varcomps</span></span></code></pre></div>
<pre><code>##  Groups   Name    Variance   Cov          
##  Name     PC1      4.5646764              
##           PC2      4.1697507  1.031       
##           PC3      5.6675581  2.132 -1.557
##  Name.1   CA.2011  6.7843697              
##  Name.2   CA.2012  0.1581563              
##  Name.3   CA.2013  0.5836402              
##  Name.4   FL.2011  0.9528070              
##  Name.5   FL.2012  0.0042999              
##  Name.6   FL.2013  0.0000000              
##  Name.7   MI.2011  4.4550266              
##  Name.8   MI.2012  2.7513777              
##  Name.9   MI.2013 13.1460692              
##  Name.10  MO.2011  0.0000000              
##  Name.11  MO.2012 11.0314401              
##  Name.12  MO.2013  0.0169141              
##  Name.13  NY.2011  3.9489097              
##  Name.14  NY.2012  0.0394018              
##  Name.15  NY.2013  7.9526228              
##  Residual          4.1133631</code></pre>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" tabindex="-1"></a>vcFA <span class="ot">&lt;-</span> vc[[<span class="dv">1</span>]] <span class="co"># G of PCs only</span></span>
<span id="cb46-2"><a href="#cb46-2" tabindex="-1"></a>vcDG <span class="ot">&lt;-</span> <span class="fu">diag</span>( <span class="fu">unlist</span>(<span class="fu">lapply</span>(vc[<span class="dv">2</span><span class="sc">:</span><span class="dv">16</span>], <span class="cf">function</span>(x){x[[<span class="dv">1</span>]]})) ) <span class="co"># G of diag model</span></span>
<span id="cb46-3"><a href="#cb46-3" tabindex="-1"></a>vcUS <span class="ot">&lt;-</span> loadings <span class="sc">%*%</span> vcFA <span class="sc">%*%</span> <span class="fu">t</span>(loadings) <span class="co"># G of unstructured model</span></span>
<span id="cb46-4"><a href="#cb46-4" tabindex="-1"></a>G <span class="ot">&lt;-</span> vcUS <span class="sc">+</span> vcDG <span class="co"># total G as the sum of both Gs</span></span>
<span id="cb46-5"><a href="#cb46-5" tabindex="-1"></a>colfunc <span class="ot">&lt;-</span> <span class="fu">colorRampPalette</span>(<span class="fu">c</span>(<span class="st">&quot;steelblue4&quot;</span>,<span class="st">&quot;springgreen&quot;</span>,<span class="st">&quot;yellow&quot;</span>))</span>
<span id="cb46-6"><a href="#cb46-6" tabindex="-1"></a>hv <span class="ot">&lt;-</span> <span class="fu">heatmap</span>(<span class="fu">cov2cor</span>(G), <span class="at">col =</span> <span class="fu">colfunc</span>(<span class="dv">100</span>), <span class="at">symm =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p><img role="img" src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAYAAAAUg66AAAAEDmlDQ1BrQ0dDb2xvclNwYWNlR2VuZXJpY1JHQgAAOI2NVV1oHFUUPpu5syskzoPUpqaSDv41lLRsUtGE2uj+ZbNt3CyTbLRBkMns3Z1pJjPj/KRpKT4UQRDBqOCT4P9bwSchaqvtiy2itFCiBIMo+ND6R6HSFwnruTOzu5O4a73L3PnmnO9+595z7t4LkLgsW5beJQIsGq4t5dPis8fmxMQ6dMF90A190C0rjpUqlSYBG+PCv9rt7yDG3tf2t/f/Z+uuUEcBiN2F2Kw4yiLiZQD+FcWyXYAEQfvICddi+AnEO2ycIOISw7UAVxieD/Cyz5mRMohfRSwoqoz+xNuIB+cj9loEB3Pw2448NaitKSLLRck2q5pOI9O9g/t/tkXda8Tbg0+PszB9FN8DuPaXKnKW4YcQn1Xk3HSIry5ps8UQ/2W5aQnxIwBdu7yFcgrxPsRjVXu8HOh0qao30cArp9SZZxDfg3h1wTzKxu5E/LUxX5wKdX5SnAzmDx4A4OIqLbB69yMesE1pKojLjVdoNsfyiPi45hZmAn3uLWdpOtfQOaVmikEs7ovj8hFWpz7EV6mel0L9Xy23FMYlPYZenAx0yDB1/PX6dledmQjikjkXCxqMJS9WtfFCyH9XtSekEF+2dH+P4tzITduTygGfv58a5VCTH5PtXD7EFZiNyUDBhHnsFTBgE0SQIA9pfFtgo6cKGuhooeilaKH41eDs38Ip+f4At1Rq/sjr6NEwQqb/I/DQqsLvaFUjvAx+eWirddAJZnAj1DFJL0mSg/gcIpPkMBkhoyCSJ8lTZIxk0TpKDjXHliJzZPO50dR5ASNSnzeLvIvod0HG/mdkmOC0z8VKnzcQ2M/Yz2vKldduXjp9bleLu0ZWn7vWc+l0JGcaai10yNrUnXLP/8Jf59ewX+c3Wgz+B34Df+vbVrc16zTMVgp9um9bxEfzPU5kPqUtVWxhs6OiWTVW+gIfywB9uXi7CGcGW/zk98k/kmvJ95IfJn/j3uQ+4c5zn3Kfcd+AyF3gLnJfcl9xH3OfR2rUee80a+6vo7EK5mmXUdyfQlrYLTwoZIU9wsPCZEtP6BWGhAlhL3p2N6sTjRdduwbHsG9kq32sgBepc+xurLPW4T9URpYGJ3ym4+8zA05u44QjST8ZIoVtu3qE7fWmdn5LPdqvgcZz8Ww8BWJ8X3w0PhQ/wnCDGd+LvlHs8dRy6bLLDuKMaZ20tZrqisPJ5ONiCq8yKhYM5cCgKOu66Lsc0aYOtZdo5QCwezI4wm9J/v0X23mlZXOfBjj8Jzv3WrY5D+CsA9D7aMs2gGfjve8ArD6mePZSeCfEYt8CONWDw8FXTxrPqx/r9Vt4biXeANh8vV7/+/16ffMD1N8AuKD/A/8leAvFY9bLAAAAOGVYSWZNTQAqAAAACAABh2kABAAAAAEAAAAaAAAAAAACoAIABAAAAAEAAAEgoAMABAAAAAEAAAEgAAAAAKtAJY0AAD9eSURBVHgB7Z0HvBTV9ccvTaoVUBAEG0VUUFFUUAERJQaxghU1BrFhFzWIaLB3jWCJiBpULAhGsYKgoihRiUEExAYEbH8wGBQsvDf/+71x1tl5M7t3d2d297055/NZdsq95975zdvDLef8Ti1HixIRBAQBQaAECNQtQZs1usl58+apRx55RO20004le076sPXWW6uNNtqoJH1YvXq1+vHHH9Xw4cNL0r40Wn0QqCUjoGhf1rhx49Tll1+u+vbtG63iHLSV2gB98cUXatmyZWrx4sU59FqKJhEBGQFF/NYZeTD6+dvf/hax5uqjbtq0aeqmm26qPh2WnpYMgdola1kaFgQEgcQjIAYo8X8CAoAgUDoExACVDntpWRBIPAJigBL/JyAACAKlQ0AMUOmwl5YFgcQjIAYo8X8CAoAgUDoExACVDntpWRBIPAJigBL/JyAACAKlQ0AMUOmwl5YFgcQjIAYo8X8CAoAgUDoExACVDntpWRBIPAJigBL/JyAACAKlQ0AMUOmwl5YFgcQjIAYo8X8CAoAgUDoExACVDntpWRBIPAJigBL/JyAACAKlQ0AMUOmwl5YFgcQjIIyIAX8CU6ZMUQ8//LDq1KlTwN3Ml6BDff/99w0ta+aSVe+uWbNGff3112rixIlVb8oVQaAGIiAGKOClfvbZZ4bPeNdddw24m/lS8+bN1bbbbqvq16+fuWDA3VWrVqnZs2cH3In/0k8//aTq1KkTSUPr169XFRUViu8ohMQttWrVUnXryp9rFHiWkw55owFvo3Xr1mrHHXdUI0eODLgb36W5c+eqOXPmxNdAiOYvv/xSbbnllpH9wDE+GI2GDRuGtJjb5crKStWjRw/1+uuv51ZRSpc9AmKAyv4Vxd/BJk2aqA033FD997//jb+xPFq4+uqrFSM0kZqHQGIN0Kabbqr4BA3rV65cqb7//nvVvn37wDdO2plHH31UDRgwIPC+XBQEBAE7BBJrgBo3bqzGjx+vWrVqVQWp//u//1Osx3Ts2FF16dLFTCdq1/5tw3Dt2rVq0KBBql69eqm669atU/fff7866aSTUtfkQBAQBDIjkFgDxMhnm222UW3btq2CULt27VLXKPPAAw+k7Yh9++23Juuod/S0//77K2+9lALfAXUxZkELvuyC/fLLL4rvICHbKFOlBg0aBN2Wa4JAtUMgsQbI9k1hLBo1aqRYJ3HFe+xew6B4R0nudf93hw4d1A8//BA49cP4YGQYlblrHuz+uPLzzz+r3XffXR1zzDHuJbVo0SIzFezfv3/qmhwIAtUFATFARX5Tu+yyi7rkkkvUAQcckLHl3XbbTW211VZq++23T5WbPn262nzzzdWKFStS19544w3Frp0YoBQkclCNEBADVKYvi6kh60mHHXZYxh6yXsU0sTrJq6++qnA5II21jbz11ltmajp58mSb4mrp0qWqc+fOqk+fPlblpVDpEBADVDrsi9ryqFGjFLt3+Pv4hUV1poWUCZLPP/9cdevWTZ199tlBt3O+9uSTT6oXX3xRMRq0kcWLF5uNgEceecSmuMIbvWfPnmKArNAqbSExQKXFv2itv/fee2aNKmjRnUXt7bbbLm1Xz9uxZcuWRbrwTYgLa1tjxozxNhPZ8QUXXGCmr5EpFEWxISAGKDZoy0sxhmfnnXdWZ5xxRmDHcPYLExbEo/JqDmtDricTATFAyXzvRXvqqVOnmvUe7zrVtGnTzHRwwoQJqX4QhLvJJpuoIUOGpK7JQc1HQAxQzX/HOT0h3t1M1/ASd2XJkiVm+uaN0sefifgs1nMyybPPPqtmzZpl3Afccv/+979N2AeGyJWPP/5YbbDBBmKAXEAS8i0GKCEv2vYx2frHofKUU05JVfnkk0+M46R3FHPPPfcYg5EqFHKANzle57feemtIif9dvu+++9Q777yTsYzcrHkIiAGqee+0oCfCWEApAhuAK95j9xr+SDhOiggChSDwW4BTIVqkriAgCAgCeSAgBigP0KSKICAIRIOAGKBocBQtgoAgkAcCYoDyAE2qCAKCQDQIiAGKBkfRIggIAnkgIAYoD9CkiiAgCESDgGzDR4OjaCkSAh9++KH6xz/+kTGSfuHChQpa3ZkzZ4b2CsfHfv36qTZt2oSWkRvxIyAGKH6MpYUIEcAbG6fFIN8ktxmMFGmRli9f7l6q8g0dCEG4J554YpV7cqF4CIgBKh7W0lIECGB44PkhKUAhsu+++5r8bYXokLqFIyBrQIVjKBoEAUEgTwTEAOUJnFQTBASBwhEQA1Q4hqJBEBAE8kRADFCewEk1QUAQKBwBWYQuHEPRUIYIsMv1zDPPmOSSQd2D44j7QTtlpKj+7rvv1PDhw4OqyrUIERADFCGYoqp8EIAEje36/fbbL7BT5FibM2eOghzNL2Qa+eijj8QA+YGJ4VwMUAygisrSIwDJPrnVvCyOtr2aPXu2GB9bsAosJwaoQAClevVGgLTb5557rmrZsmXqQZh+MQoii60rEPOvX78+cMrmlpHv3BFIrAHCC9ab9jgMOjxqbVIu25aD9zgoL7y//ajL1atXLzAdtL9d23J169r96VAOndnEthzYgU02sS0HBS1JDMePH59SyRoQudC6dOmSuoZ39bXXXps6l4NoEKjlaIlGlWgRBAQBQSA3BGQbPje8pLQgUKMQKDWvt904ukZA3sTqKTodNcyqnBQSBEBgwaTrswKxww47KDKL+IVI/E8//dS4Axx55JGKdSab6T7l7rjjDvXUU0+ppUuXqqOOOkqddtppJvEkbaxZs0ZdfPHFasqUKSah5DnnnKPOP/98f/OKVNdXXHFFlb6RSunBBx9UM2bMUMTMDR48WA0cOLBK/SguJMgARQGX6BAEckeAxeszzzxTnXzyyWmV3bUsVkEoYyt/+9vfzJoVqZHY7bvqqqvU/vvvrxYvXmzyuV166aVqwYIFavr06eqHH35Q5Hpr1aqVGjRoUKoJDAxZcrnuFRJEUu4vf/mLQv/LL7+sjj32WLNuecQRR3iLRnIsBigSGEWJIJAZgS233FLtuuuumQtZ3n344YfVsGHDVK9evUyNm266SbGbB08Srgd33323ev/999VOO+1k7mP8xowZkzJAUJC8+OKL5vzNN99MaxW3he7du6tTTz3VXD/++OMN88Bzzz2n4jBAsgaUBr+cCALlj8Brr71mDJDbUwjYGEHhSsCUrmHDhmZnz72/1157qXnz5rmnqnfv3mrRokWqb9++qWvuwXnnnadeeeUV91RVVlaq+fPnp7kppG5GcFCUERBpfkl4Z7t1G8FzGRVstQO0iCBQagQYoWA4vIKntn8K5L1vc8x6EFOugw8+2Bidxx57TDVt2jStKr8/fJvWrVtnjNMf/vCHtPuZTm6//XZTl+laHFI0A4SfhTe1bxwP49dp4+fjryPngkAcCGy77bapKZOrv1GjRu5hXt/ff/+9WYDGuDz55JNGx48//ljlP3obv7OgDtx2221mkXry5MkFG8og/VwrigFi5IOlb9u2bVg/5LogUKMR6Nmzp9mZiuoh8dQ+6KCDzMzi1VdfVRtvvLFRzVrTt99+m9YM51tssYUZ/aTdyHBy4YUXmnUlFqH33nvvDCULuyVrQIXhJ7UFgaIj8MUXX5gg26233trsUrnGh460a9dO4cm9bNmyVL9Y/7GdfbAjN2TIEDOieuONN2I1PnSwKCOgFBJyIAgIAqEIfPDBB2l+QEydOnXqZNYxie53d6b4Zn1z1KhRaT48hJVgaA488EA1evRowwbwzTffmF0xtupthC1+dtmY0lVUVCj6hGy00UaxzGDEANm8FSkjCBQBgV122SWtlc0220ytWrVKvf766+r00083BgiakOeff96U82/r41h43HHHqbFjxyocG5s1a2bi8NhKx5fHRm655RbjEInvkFcOPfRQ9fTTT3svRXJclFiw9u3bK/wIGB6WSjoddalV0wsmjbEqt8PR0XpMO/WtmlWVDe1C9+qsrmWlsLKRnb66q+1m6+v2/sWq3bpL61iVW79FpVW5OqvsntfJHsdq1Z5b6JObbnQPy+77q6++MjtiNsHApep8LCMgXMA/++wz1bp1a/NcJIlj0ayUBqhUAEu7gkCpEGjRokWpmrZuNxYDxBwSt3A3eRzbhQwlRQQBQUAQ8CIQiwFi4Yz56ciRI01bTME6duzobVeOBQFBQBAo3i4Y5E6s0NtE+0b1XiAJe/fdd6NSJ3oEAUEgYgRiGQEF9RH/AtzRC/X+DNIddq2Yxi6sD3JdEBAEwhEomgHCGDA1a9KkSXhv5I4gIAgkCgG7vdVEQSIPKwgIAsVCQAxQsZCWdgQBQaAKAkUxQMSurF27tkqQXJXeyAVBQBAoGQKEXmTKUREHf3RR1oAeffRRw75GPEm5i62H88LH7TymOww5y+qRa/1s58lrpUwXqmhm5+Fsq2/dNnaUofUW2nk4W/evwq6HFS2ifV5l59Bt1TlCInBLIbLcSxFDvBXMhtCgEl5BuANOvP3790/TO2nSJMPLTHBo165d0+5FxQ9Nhtg99tjD9ANWRb+E8Uf7y+V6XpQREHEltvmmcn0AKS8IlDsCjCzgZ4Zj2SteLmgIxYjXIuYLfh9XcOA966yzDOmY3/hQxuWHvuGGG9Q777yjSDkNP/R//vMfo8LLDw1Z2fXXX6+eeOIJV735JuD0d7/7nTGEaTd+PYE/mgj5OKQoBiiOjotOQaA6IUB2CTJVLFmyJLTbZLrAKF100UWpMmeffbaJQr/ssstS17wHXn5oouHhh169erXhhyb8CX7oO++80/BD77nnnoYcH35oV8iswciHCPoggT+afnsJ7YPK5XtNDFC+yEk9QSAHBJhqEe1+yimnhK6zQJ167733qvvvv18x3XrhhRdMBPqECROqsBy6TRfKD73JJpuot99+W1133XWuyrTvTPzRaQXzPCnKGhB9g4+WoWG+9JD5PB+e0LDFiQgCpUYAPzgccYkIYFRCpoogYf2HUQf3Salz4403puWoD6rjXsuHH7pPnz6mOnWDJBf+6KD62a4VzQBh1YmGL6Z3cjHbyga03BcEtt9+e5NfnikNay5hAhF8hw4dTDA36z82Ejc/tE0f8iljDBCUjeQUguIxCkFf8+bN01SddNJJaedyIggkEQGylLLuwlSMbKZBwrQIEnsWnb27ZkFluRY3P3RYu1FcNwaILTZW091EZoUqJikaAIoIAoJAOgIYFKZinTt3Njm30u/mfoaPHdMo0j/zOyYnmCtefmjSQCO58EO7euL8NgYIw0OSMoxQFHL55Zcr1l9EBAFBoCoCpFNm0ffcc89V0K7mIuS5KxY/dC79yres7ILli5zUEwQKQIDt9f322y9nDfBDu1M3lx+aGQeOjoyq3I/LGw0/NJQ0OEOyAI6zoy0/dM6dy6OC4YRmm2/atGmRj4BcQrI8+hV5FVtOaMfOkVdVbmjnefvRuLFWz9LecrHRqWvXbkVTu3JOY7tylTvZuSTXnWW3r1HZ3LLdrew4odUWdvpU8GZP1Xdk+XewtOdtVeuW4ZVy5Ye2+2spQ0ClS4KAIGCPQLnyQ8sUzP4dSklBQBCIGAExQBEDKuoEAUHAHoGiTsHIUY2DlT8Bm313cyu5wQYbmHxkudWS0oKAIFAsBIpqgPBMxr38kksuKcrzFTPsoygPJI0IAjUMgaIaIAxC3bp11QEHHFDDYJTHEQQEgXwQkDWgfFCTOoKAIBAJAmKAIoFRlAgCgkA+CIgBygc1qSMIFAmBNWvWpNgNi9SkaSYbP3RUfSnqGlBUnY5Tj2MZwmbL4Wzr4bxYu8zbSLsLgnlk/HXrrLbjmK6o568ZfF5nrt2fSkUHO8/l2sst/++za1apFXbPa50L2LbdYLjSrm644Ybqxx9/NCERhEN4BbrTU089VRElf8stt6RuTZ482dCwfvLJJ4bChlCLI4880lxLFfIdZOOHpjhcRPfdd5/67LPPVL9+/cwx/fNKGD/0s88+q+jvjBkzFAyPgwcPVgMHDvRWzfnY8q8gN71YbaJ0586dW+UTB7N+br2T0oJAcRFgNEFwtp+LmV5MnDjRUG5QxhUMEaRkcAF9+eWX5reEgbr55pszGqBs/NAzZ85UV155pdFBHNl///tf5SccC+OHhjgfWlaM1uLFi9XRRx9tYsowlIVIhHb+t27Q2dmzZ6s5c+b8dlEfYXz4n0BEEEgaAjAdkgXjmmuuST06PD7wcHXv3j11jdHHqFGjDI8znEGuMNqArwvqjX322adK5gzKefmhOYcfGuoP2jjooIPUVVddpYYOHZrid8agEby6YsUK1apVK8NTdPzxxxs2xg8//BAVKcFQ0k9GawjlyHbz3HPPqSOOOCJVLteDWAwQnQ0SRkY8qIggkDQEDjnkEPXMM88oItddR1zS7ZANAxJ5V6ZOnWo4ffwjE+4z7enWrZuZAvlT93AffmivLFy4UK1fv161bNnSXGY65yW8J1V6o0aN1IIFC8zv0uWHhlvottvSg2zPO+88xceVyspKNX/+fBOF717L5zsWA5SpI8xTyTvUtm3bTMUiuYcn9OOPPx6JLlEiCBSCQJMmTYyxYRTkGiDS5AwfPtysy7i6mRrBFxTGhMhvBxL5bOLnhyZdDyOdpk2bplWFj4hpHpKNH9pbEdpY0gedccYZ3ss5HxfdANFD0ocUg6JVPKFz/nuQCjEiwLrJiBEjzDQMY8DIg6kRC8OuMLJo0KCBe1rlm3uZspdSIYgfmuUPdOMI7JV8eNMZHV1xxRWK9Z9CZzTpvfH2LKZjLDvk3IcddlhMLYhaQaA8Efj9739vFn2ZhrGTdPjhh5uEnd7ess7CqB0jEzQKghFx77339lZJOw7jh27cuLEiMzHxmF4hgeE222zjvZTx+MILLzTrSmR5zdSPjEo8N2PZBfPol0NBQBD4FQHWW1gLYhrGjlgQMyEsiexOPfTQQ1VwY2EYpsNevXpVuccFdp6pz2I1BmLjjTdOK9e+fXvDCe1eXLZsmVl/sjFAGESyo9J3cpZFYXzohxgg923ItyBQBASYhrEzxW5Xz549q7S45ZZbmq15tt3JYLpy5UoTwI0PDtM1FoJJdY6wroNPz+eff27O2aFiisYuGgvObKnzcRe5SfvMdI9dagzK6NGjzSisdevWpn6mf9jiZ5cNildcBlzdS5cuzVQt672iTsHYggc08mSLCAJJRIB8YDBCsMUetv7C8sT48eNNllQWqVm/2XHHHc2CrzdFM3rYVmdUwu/K5YHGadErZMs47rjjjG8RC9iMeHA+3HnnnRV0zDaCbxIL267xc+vAMf3000+7pzl/x8IJHdYLDBBbiZtvvnnReXpsOaHXW3Iphz2j/7rTwJKr2JKD+ONb7/I3EXhu6zFduUFg9bwvru+23q5ufTtcanf/zUEvk+LKf1gC2CSTFs89S+7opfunb1d7NERySEbhtWvXVtm9KkQ5hguDRSroUktRR0AMD5n3sgMgIggIAtkRIM+XN9dX9hrZS7AgzaccRNaAyuEtSB8EgYQiIAYooS9eHlsQKAcEijoF44HxY2AbD3+FuAVP6HHjxsXdjOgXBASBPBEougFiFZ1tPxvfgzyfKVVNPKFTUMiBIFCWCBTdABFEFxRIV5boSKcEAUEgVgRkDShWeEW5ICAIZEJADFAmdOSeICAIxIqAMUDz5s1LixGJtUVRLggIAmWBQDbe52Kwl5o1IILX+IgoZc2l3MzOk7fC0rPatl1bD2dbj+mtJ5xm9dpr9f3FqlzdMXak2hXb23FHV86zalbV7mrnMe18X8tOYYSlcuGEJlqA8sR+QX/ql7h5n932CN+AcoOYsjjFjIAI0+cjIggIAtEjkCsnNCyGYZw/cfI+u08O8TyR78UQWQMqBsrSRuIRcDmhvUAEcUJ77wcde3mfIfaD95lod3ifES/v8xZbbGGI7AkWdcOfnnrqKbXHHnuoAw88MEi9CVi9+OKLU7zRgYUivCgGKEIwRZUgEIYAPEDLly83nNBuGZcTOpe4LHifhw0b5qpQQbzPXq4eL+8zlVze5+uuuy6lw3vQu3dv4yzct29f7+XYjovuB+R9EnhL3nvvvdj4oevVq6fuuOMOb5NyLAiUBAFbTuhcOhcH73MQGX4ufcq1bEkNEAxucKLASxKH+Plv42hDdAoCtgjYcELb6oqb99m2H4WWK6kBgv2N7BiFMusXCoLUFwSKgYANJ7RNP+LmfbbpQ1RlZA0oKiRFjyCQBQEbTugsKmLlfc7Wdhz3xQDFgaroFARCEMjGCe2vVkzeZ3/bxTgv6RSsGA8obQgC5YSADSe0t7/F5H32tlusYzFAPqQrG9l5OPuqhZ46je30VdQLVZF2o9Z3dp68th7OSwbfm6Y/7GSblaeE3Uq7XnGMnce0s9TuOersaccx3bDBd2n9CD1pFnon7UZtZeepnVYp5AROZ1fq169vMoq653yTX90Vf+JBeJu9ToneY7eO95udXzJlkLk0E+8z/cik64QTTlB84pZYDRApX9l+DBNeDICLCAKCQLQIlBPvc6Yni80A4aPALhdxLWGCAdp2223V1VdfHVZErgsCgkANRiA2AwQbIX44ZHkMExwRGTKKCAKCQDIRiM0A5QMndK3Et+Timp6pHYzbNddck6mI3BMEBIESIlBWBoiwjHbt2qnmzZtHAol4QkcCoygRBGJDoKwMECv+pKwlDa2IICAI1HwExBGx5r9jeUJBoGwREANUtq9GOiYI1HwESjoF+/zzz9WyZcsUW/bIkiVLDAWkTMFq/h9eUp8Q579Vq1apZs0sPSKLCBTMjbBT1Kpl5yQaRddKaoC6detmHBEbNmxonoWHL3Uywbqr7QaF67ax89Ct3MmOq7jO3GhfhS2Hs62H8+fNxlv9vXUYfpZVuZ8u+9GqXC1LP9Wh6m0rfQvVplbloi700UcfqREjRigIxTBAUNCw64srChl8vfKf//zH7Abvtttu6vXXX/feqnKcjSN6zZo1CobDKVOmKH5n55xzjjr//POr6Pn3v/9tmBKff/55RbuuwE09ZswYNXv2bLXrrruqsWPHRkqfE+1fvdtry++zzz47reTEiROLkjE1rVE5EQRiRoCRfp8+fQwN6jPPPKN22mknNXPmzBSzod8RF0J4jMA777yjPvjgg4w/eJcj+p577lHbbbedoWTdf//91eLFixWbOpdeeqlasGCBmj59uiKuDKPXqlWrNMpV2jj22GPV119/nYYEkQxcv/vuuxV9uvnmm1XPnj1NRH5UEQx2/92ndUtOBAFBIBcETj75ZNWxY0c1btw41b17d5MA4tBDDzWjCX7cLl+zq3P8+PHGQBC4yv1MkokjGt4g6t95553G6O25557qzDPPNCMaV2cmjuh3331XwZA4ePBgM2UcPny4YnT2z3/+061e8HdJR0AF914UCAJljgC5tSCM54fOEoNXGI1gJLzX33//fZOj74gjjlAtW7Y0BuDGG28MjalkSucVL0f0p59+aqZdnTt3ThXZa6+91G233ZY6dzmid9hhh7TrFIDHmg9SWVmpGGURWsUILiopKwP07bffmofcfPPNI3k+HBH/9Kc/RaJLlAgC+SCAQSHXF+TwQeI1Pty///77FYTwxFHy42d9iOnPaadlz9/m54h+7LHHVNOmTdOaZVr23XffqXXr1hnjxNQQcTeC0gr/esKIhz7x+3zxxRczxncG1c90Ld0kZypZhHs9evQwgPO/RhQf8iuJCAKlRMD1xs/0A3f7R5lHH31UHXzwwYr1FxaroXHNNg2jPhzRTOswLg8++KBRieFz2zcX9D/5bPK0adNGvfTSS2YRfeDAgVkXxt22bL7LagT05JNP2vRZyggC1QYBpiuMYlgI7tChQ1q/2ZJ/4IEHzOJ069atFfm7WGO56KKLzMctDK/PW2+9pbzpdtx7fIdxRDOKYtTiFc7JF+buPHvvhR0ziuLTtWtX0w/6vN9++4UVz+l6WY2Acuq5FBYEqgECBERjOCAJYx3FK0yRWBRmOoQw/TrppJPMdIjRkPvZZZddQkdBZJbBGJBa/eWXX1Ybb7xxqgniKmGjwNfOlXnz5lnvNDPyuuyyy9yq5hv9K1euTLtWyIkYoELQk7qCgAUCjBjmzJmjTj/9dDV//nyTyZRpEv45jHYwFBiJV155xew4+VUSH8nsgCnZokWLjDFzy5x66qnGlw5/IvK4s6XOh2yp22yzjRldjR492rAfss2OUcEXyEbYNSPzKn1ntPbCCy+YnPWDBg2yqW5VRgyQFUxSSBDIHwEMwd///nf11VdfqX333VdtttlmaujQoWZ9xx1hYJCYMvXq1atKQ8cff7wxAJTBMdFdkMa5EcdBFrpxEmS3y/1wHcFxkO10PK+7dOli1onw7bERfJHYgcONgCnYkUceafyM2JaPSmppy+ZMmDBBTZs2TeHUFJWwAMw8k8XkcpBOR11q1Y0fu9ktXNdbWMdK3/pW6cPusEoVHezK2XJM151r1z9bDuf6t9QP63ra9Y8eGJt2Hnay9fwhYbfSrp+80/S087CTJ9Vv3rthZbjeRqU724WV1X+5YbfSrk9WM9LOs50wDcPrGGNTTDI+jB9GJN826TNuAf5F7WzPm+2+WYRmuMZcEiMUlWCA/HPeqHSLHkGguiLAtjvJOIstLVq0KKhJiALjEGOA2K5jDsp8LyohsE0MUFRoih5BoGYiYAwQLtZ8ohR3ChalTtElCAgCNQsBY4DK4ZGmTp1qVtiJmYlKmK/6A16j0i16BAFBoHAEysYAzZ07V82aNSsyQnqgyXfBrXBYRYMgIAjYIFA2Boityt13313deuutNv2WMoKAIFADEBA/oBrwEuURBIHqioAYoOr65qTfgkANQKBkU7BXX33VuJe7NAX4IOHsJCIICALJQaBkBohFZ7hFXAJsHCEzpXEu1iupu9TSg7iZY9WlyuZ25WovtxuMVnS289Su2N7Ss3qpHQG5LYezrYfzkp3GWeHXQx1gVa6upefy12oTK32ant2qXK6FiKmqrqT0RDVEvbETD8oWb4XoXaJ8Ibzmw3Z5GN2AhTopIgiUNQLEbRFLRdZfPsRsjRw5UkG14RcoOZo0aWJFeUHEPPFaBI7i7Txs2DATjOrqhJT+jDPOMPfY6PGyIbpl+Gb2AVd0GN0qpGiwJkYtJTNAUT+I6BMEyhUBl5QeKgtI6SENu+qqq9RDDz2kiFT3Cz92Lym9/7733CWlv+GGGwyJPQYNUnqMGOIlpYf+4/rrr1dPPPGEV4UxWPBP+0np3UIEwQ4ZMsQ9jfRbDFCkcIoyQaAqAtWVlJ4nOfHEEw1tSJQUHF6ESrYG5O0Ex1jfjz/+OI3rxF8m13M8oWH1FxEESoUA6ybVlZQezHr37q1uv/12Q/vx5ptvRg5j2Rgg2PmhriQXUlTCgpkYoKjQFD35IFDdSenj/v2UjQFijhnXPDOfPxypIwhEgYDLn8NicTahDKT0V155pSGlp7xLSu+SkIXpgJT+qKOOMutLLrd6VKT0YW1GcV3WgKJAUXQIAiEIeEnp/UXYkicJ4fLly80tLyk9u8R8WDj+17/+Zcjg/fXdc0jp4YUm8yn+deyyIVGR0rvtxPEtBigOVEWnIPArAiwDVFdS+mK8RDFAxUBZ2kg0AtWVlL4YL61s1oCK8bA2bazfws6DWFXYaNMpbbey1Gf5Jmp3t2u4cp5d/+rsaedZXauBnb7BaqZVQVsP5zeVHSf0M5Yezq3Vj1b9i7KQS0qPwyCk9PgBsTZEtgtbUnoIA/HHITUy2TXIhuGS0tNXSOm9gi/RcccdZ0jpcYCElJ7RGAT3tqT0Xn1xHRtS+jiUs/3YoEED9ec//zlQPYnWFi9ebLbeAwtEfNGalL6z3Q/c1gBVdLHUZ2uA9rUzGJXz7EJKSmWApqntrd5wuRug3VRuObKqKym91cvKo5Dln33umonxItVy2Oq/kNbnjqnUqP4IVFdS+riQj80AMcQkh1GYTJ48WTFMFBEEBIHkIhCbAYoL0g8//FBNnDhR7bjjjlmbqFOnjorLhTxr41JAEBAEsiJQ7QwQbu3k2e7Tp0/Wh8OzWgxQVpikgCBQMgSqnQHCOYvRDx6jIoKAIFC9ERA/oOr9/qT3gkC1RkAMULV+fdJ5QaB6IyAGqHq/P+m9IFCtEah2a0Bxo11nVS2rJipa2HE9qy0sy62wa7fyH3YOhrW72jlANmzwndXzDlVvW5X7q9rLqpwth7Oth/MAtdqq3TWWXM8bKEsPdqtW8y8EpSo+c5tuumn+SvKoWVGhWbFr105xtuehwqqKjICsYJJCgkD+CBA+QRgEUe1+IbyCexdeeGHaLfzk2rdvr6BxJbp9jz32MHSqaYV8J9n4oSl+9913G7pX+LeOOeYYhYHzSxg/9LPPPqsOOuggEw5C9P0HH3zgr5rzuRignCGTCoJAbggwmqhfv34VLma04NNG1ABlXLnlllsMFepZZ51leIHIGHPOOeeom2++2XA8u+X839n4oWfOnGm4huCJJo6MLDR+wjGMShA/9JdffmliyE444QQFx3X37t1Vz549FZxDhUjJpmBLly5V8+bNUxdccEFO/V+4cKHCGVFEEKhOCPTv39/kwbvmmmtS3YbHB782fsyuMPoYNWqUuvPOO02wqnt98ODBChcU/N/22WcfhT6/PPzwwyYrRq9evcytm266SRGJTxuMXCDCHzp0aMo3DoNGdo4VK1aYjBhPPfWUCVY988wzq/zG3n33XWOs6AdCcCxE+GTRKCSbTckMEA+OBd1qq63MA9n+s3LlSkPdGlSeXGOzZs1S2223nbmNJzTWXEQQKDUChxxyiMmIAUUr6aiQSZMmqYMPPlitXv3b+tXUqVNVw4YNq4xMKE8kfbdu3dSMGTMCDdBrr71GsZTwnzXrRy1btjTXPvnkE3XRRRel7pMUtFGjRmrBggXGADEte/vtt036HX/6HvrPByGg9p577jFTMQjXCpGSGSAsuY03s//hGEa6DHL+e6Q8wUualCYIntBigPwoyXkpECDPF8YGulTXAMF2yEiCdRlXmBrxH6ibsNO97n7zt42RyCasBzHVok3+syddDyOdpk2bplXdbLPNUvSv7u8xLICciox4+vbtq7799luTWJT1rUKkRq0BdezY0VBTsljGhyGliCBQLggcffTRqXUgjAEjD6ZGXmF0AY1NmHAPKtdMAj/0oYceaniHWORGoMdBt8tRbS7qf9jpykXatGmjXnrpJTVixAg1cODAjAHnNnpza91Go5QRBASBQAQgmGcxl2nY448/rg4//PAqqY5ZD1q0aFGokeFepjWXMH7oxo0bq4022siMXLydI4EhhGm2wgiqa9eu6uqrr1a77767WWOyrRtUTgxQECpyTRCIAQHWW1hHYRpGdtIgZkK2t9mdImuqX9h8ef7555W7yOy/z24Z9Vmsfvnll80WvrcM2/ps/LiybNkys/5kY4CYJrrsjW59XARYky1ExAAVgp7UFQRyRIBpGDtT7HaxCeMXMlmwNc+2+5gxY8wPnGwXrg/OeeedpwYMGGCqsa7Dmifb4gg0rUzR2EVjwZktdT7uIjdUrhgSkoAyjSMtNKOw1q1bm/qZ/iH3PLtqc+bMMXVfeOEF06dC2SZKtgid6WFLec/ZIOLWs6eD+l+Dtm+iiV3/nO9r2RVsZldsobLzxG2jvrZS+HXEHM62Hs4bWno42742q4f1FGJTBIMCH3TY+sthhx1m0vXce++9ZpGa9RsYIM4444y0UQh62FZnRIUxYnSEhPFDk2aZBWxGPCwe77zzzmrChAme3oUfsvgNpzVppjFg+P+wre9uy4fXzHzH9s8+sxa5KwgIAqEIrF27NnUPh0RI6b3y3HPPeU/NMUkG+axbt05R3797RSHCM7wL0t7jKgr1BTyuGTGRahmDFRbeQR+DdDH64sPoja19/4J2UJvZrlU7A0T+ePx98Inwy5IlSwyw/utyLghUVwTwCeITpbAgzSdfydV3L1M71c4A9evXz8xzt9122yrPhR8Qc1QRQUAQqB4IVDsDhB8Cc9kgwUGR4SHCQh+W+oADDggqKtcEAUGgDBCodgbIFrNzzz3XeICKAbJFTMoJAsVHoMZuw7NINn78+OIjKi0KAoKANQI11gBZIyAFBQFBoGQIiAEqGfTSsCAgCNQoA4QLO7Ews2fPNr4WnIsIAoJA+SJQoxahcfCCzgCKAwwRLuoEzMUiv1hqrWNZzvZNROyiW9vSM9jyKVRDZQeMZhy2VWlVzpbD2Ra++latVv9CeFnjoFgqifavoFRP8Wu7GB624d988021/fbbqy5dupS4R9K8IKAMwRc/cv/HJc7Df4170GXYSDbuZ3ieCdto0aKFCbvwk4u5bTzyyCOmb+65++3GnRGuERX3s6vb/12jDJD/4eRcECgHBGAlhOYUalTvB8ODEPZAGVvJxv0MERlcQ9OnT1eQnl1//fUpHiK3jQc1T9CQIUPc09R3XNzPqQZ8B2KAfIDIqSAQBwJEuRMk6v0QYJqPeLmfcbYlSp2Id4wbSw9EvMMpDV0qUewYPyLrXcGR9+KLL05xQ7vX+fZyPzdr1swsZ8AZBBNiHFJjDRDDVCGvj+NPRnSWGgG4n4cNG5bqhpf7+dNPPzWxY9CwurLXXnul8QD17t3bkJ5BreoX+IowXkiU3M/+dtxz26VPt3y1+WZIe+211xq+k2rTaelojUWA0CA/aTyR6a1atSromf3cz0y5/JHzRL2zQUNkfRjhvb8TUXM/+/W75zXWAIUR17sPLt+CQDERIHjaz2QIQ2IhAvczlB0YFziBEHh6/DQZZIfJVVzu5ylTphjuZxamWZCOWmqsAYoaKNEnCBSCAOyHrLtEJaz1QGgPrcarr76aol9lrYmMFV7hfIsttsiJ1oNRlMv//NZbb5ng7jgMUI1dA/K+ADkWBGoSApm4n9u1a2c4peF7dgUeaBveZ8rHxf3s9sX/LQbIj4icCwIlQsDlcHa/2UpHyITBepErmbifMTQHHnig4Xtmex/6VIwKHNM2Ehf3c2jbupMigoAgECMC2inW0b44oS08/fTTJPqq8tFJA00dzQ3t6ESF5lgboyrl3LrasdCU0YT0jnbCdaivp16OTn8e2Lbmg3a0M2SVe9px0dE59hy9eO3oRWvnuuuuq1ImqgvmqUKtk9wQBASBaovAV199ZdZx8g21iJL7OQxEMUBhyMh1QUAQiB0BWQOKHWJpQBAQBMIQEAMUhoxcFwQEgdgREAMUO8TSgCAgCIQhIAYoDBm5LggIArEjIAYodoilAUFAEAhDQAxQGDJyXRAQBGJHQAzQrxATPa+dq2IHPN8GoEaoqKjIt3ps9cDthx9+iE1/oYqj7l/U+gp9vupePzEG6PbbbzfUB82bNzcETdBWegXypkmTJnkvFfUYMqmuXbuqJk2aqP333z+Nv4WOXHnllerYY48tap/8jU2dOlWdf/75KZxGjBihwHOjjTZSAwYMUCtXrvRXKep51P2LWl9RwagmjSUiGh5yen7AF110kYkaJjame/fu6vXXX1dwpeQqMMShM5ug2zYI8KyzzlIdOnRQ2m3eUCsQk/Pcc88ZY5StHf/9OPr317/+1cQT9ejRQ40dO1a99NJLatq0aUq76RtMb7nlFkPxqcMK/N2pcs5I05Zhb7fddqtSP+hClP1Df9T6gvos1zQCUcV0lLOea665xtH/c6e6uHbtWufggw92NFOcozlVzHX943eeeOKJVJlMB5pr16ldu3ZoTA6w8hk4cGAmNal7ixcvdjT/iqOnWalrd911l6NJwR09MjLXLr/8cmt9UfePDhAbpI2O6cu//vUv83z333+/Oeefjz/+2GCiUyGlroUd6GmMiVFyccr0HabDfz3K/qE7an3+/sr5/xBIxBQM7hT4UFyBFc6dbmkjkRMhODr69Omj+J8eQim4WCCBCvqQdcBGmLrAv6sDDlPFyWpwySWXKG0olf5xp67bHETdP9oEw5133tk0D90nIzT4jV1p27atGU3C0JdNIMh65513FJzITOmCsHOvZdPl3o+yf+iMWp/bT/n2IZAESzxr1iwT9fv222+nPa7+4TudOnVyBg0a5Gy99dbWIyBXyejRo50ddtjB0YvD7qW8vhn5aCIpR9N2Opo2M02H5v519DTOjH5sR1Sugqj6h74//vGPjiYzd3SAo6s+9a0XoR1GaJoGInXN5kDz1Dj169d39HTMpnjGMlH3L2p9GTuf4JuJmIJhIPgh6xGL8/LLL6e97hUrVhjqAm2XczZAP//8s3PDDTc46ChUHn/8cWNodKqUNFUYJ70+lNOUzlUQZf+geMDAaNJyV33qu1+/fk779u0dDH2uokeiDlPGQiXq/kWtr9Dnq6n1ExUNDzWlXrtRm2yySdo4kK1Vci2RRdWbTSCtUBFO2GaHAkGPxqq0pn/cCp7rUu+EffPNN2rzzTdP6x8EWkynwLbUEnX/otZXanzKrf1EGaByA1/6IwgkHYFEbMN/9tlnJnFbtpdN5gL/6CioTtTb3OWuTw//I902T5q+oL8hufYrAjV1bul9rtNPPz3rlrmGw3oNKOpt7nLXF/W2edL0ef8W5TgdgUQsQvMHf9pppxnfDj3acPQWb+Anl90sndfbLGrrbfhAXbTBIrCtlLs+nWXB0es8xp8qDD+u20rS9NnikrRyiTBAvNRffvnF0X4rzsiRIyN7x1Fuc9OpctcX5bY5z5s0fTyzSDoCdXSIwpVJmI+yQ7PvvvsaB7Nu3bpF8siEc+j/9RVrR9pruWCd5a4PZ07t98R/WuaZC33gpOkrFK+aWF92wWriW5VnEgSqCQKld9woAVD425BPO0pBH3qjknLXpz22FZ+oJGn6osKtuutJjAGaP3++0qEEqkWLFmqDDTYw2+1Mm3A8JEo+V04bHXRpsk9qD2DVoEEDo49vnehNHXrooWrhwoU5/W2Uuz4eRgfrKvKD46pAHBwf8Nx7773VxIkTc3reJOrLGaAEVEjEFOzDDz806z8YBh1KYAJTGzdubEZBeB4TNKoj0o3RwDhlEzyn4ezRu1xKxwwpnfnS/ChZD1q1apXS2+rqwQcfVH//+99Vz549s6kzwbDlrI8H0Nkylc7uabiUNIuA0lk3jecz3uXkHofiBHyh5bCRpOmzwSSRZdLXpGvm2cUXX+xceOGFGR+ub9++zvPPP5+xjHtTR3I7pNvNtO2s+WScP/zhD26VjN/lro/OE6zrD+b1PhRBqqQBtnU9SJo+L1Zy/BsCiZiCaf4fxYgnk0AcpgMQMxVJ3WO9Qkdxq0wpb9FnS6NR7vp4cPqYCUPiw7SrgxkBpoDKcJA0fRmgSPat32xRzT165ZVXHE0b6uDs5/8fGioJiLUg/9LrRFYg4Niof3COpiR1NG9MWh2i19944w1HU7w6V199ddq9sJNy10e/8SbX6z8OZGR+Wbp0qaHr6NKli/9W6HnS9IUCkfAbiXFEfOihhxy9buHUrVvXTBWYQmk+Y3OuF44daCFyER0B7uyxxx6OJhFz9KKs4Rtq1aqVo8nOHD1ScDS1qoNhsZVy1wfT4ZlnnmmeF1qTrbbaytCHbLzxxubaQQcd5EBhYStJ02eLS9LKJWIR2h3jMhX79NNP1ZIlS9TXX39tCNW10VDwDudDJaFHO0bX55ofWocWKJgW0QdzoE1Qq9sv97vc9dFPmAJdDFl010RqZhEeZ8x8JGn68sGoJtdJlAHK9CLZ2YIq1EuLmql8tnsYE/2/mdGZrazN/XLXxzOwBpRpXczmOb1lkqbP++xJOU7EIjQvM+q0PFGn0Sl3fThGQoZGCh6yd2j62LTfCKRkNi4MbqWk6XOfW77TEUiEAWKKRMgbRO+jRo1SOoLdpOWBhydfcdPo3HvvveZHCUn7jBkz8lWnyl0fz7lgwQL1l7/8RfXv39+QyUMon68kTV++ONX4eklY9Io6LU/UaXTKXR9/I+xwzZ49O/Xnwo6h9oI2EfxcJLJd/1hS97MdJE1fNjySej8RjIgsdBIy4IqblgfvY9Ly6O1595bVd1gaHbyCSaOjf6hWetxC5a6PfuLh7U1tBAc0iRPx9OY64Ri5SNL05YJNosomwfJGnZZHLwhHmkan3PXxN3LOOec4J5xwQpUMIGQZYVseT3P9w7H+c0qaPmtgElbQ/i+mGgMD02HUaXmiTqNT7vqYJhKugs+TX3TMm3H0zMUAJU2fHzM5/x8CiTBA7svWw34HSla/wJaIN3SQl6+/rPccR0O9wO29lDrWeeedRx99NHVuc1Du+niGMGfDL7/8MrUeZPOsbpmk6XOfW77/h4D4ASVqwi0PKwiUFwKJ2IbPBjke0kcffbTS0d7Zilrf1ztvhsLCukKWguWuT8fbGQyzPIb17aTpswamhhVMxC5YtnemB4OKHSz4faKSNWvWKLyXo5Jy1/fTTz8ZDKN63qTpiwq36qZHpmC/vjFGP8Qz+dMOB71QDFUuXr9BOmriNWLsgtJK5/usUevLtx9B9WDQnDNnjiGmC7ov1+wQkCnYrzidfPLJ6rXXXrNCDf8XykOjGofALb1o0SLFqKfchVHe1KlT1e9//3tDR1tof6PSt2LFCgXbZVxCtt0+ffrEpT4xemUKlserJtr90ksvNXzSUK/26tUrDy1Vq/A/Knrfffdd9f3335tAVs0rpI4//ng1fPjwqhVCrsydO1fdeeedIXd/u6zpRAzF6m9X7I9gExg3bpzSzI+GCYCUR+PHj7dX4CsZhT7+Q9DuFiYkBgOEtG7d2rwfQkg23XRTX6tyWmoExADl8QbggNZb9iau7MADD1Rnn322uvbaaw1LYh7qTBXt0KeOOOIIM5IYO3asofRYvXq1WRhH9xdffKFuvfVWq2h9poh4GmeTfEZYxNHB/zxlyhSluYAUfdQ+Papdu3bZmgu8H5U+DFi/fv1MRL6m4FU6CaVJFsBIEp7q3r17q2nTphkKlsCOyMXSICD+CP9DQEd4W+eG92Kmo9idXXbZxaQtfumllxzO3U+Yj4u3Psf4/0Dsdd111/lvmXN8jXAAfPrppwPvF+PiHXfcYVJbQ7imdwydF154wdFTHEevheXVfNT64N/WI1FHR9lX6Q+OqBCm6ZFklXv5Xsg19i3fdmp6vUSMgN566y0zTchk4vMZDaCPaQzpavbZZx+l/8jTmiDOjHvZhCmXZmpU/M8dJCzsMsoi2p7ME9kEagztWJmtmNKG06xlZS2oC+hQCzOKIG7OHe1o50ObqoFlotZHJhKmgNCF+AWyuZtvvlkxWrUV3pv+8YcWh4BOpHAEEmGAmDI89thjWdHKlRURwqwbbrhB4aODASLVjJcJkaBXG2EHjlQ3mdon+JN1DBvRIy+lRxhmurH77ruHTtu8waXZ9GLQ+HTs2FH16NHDGC6eOV+JUh/TUxacSW0dJvRbZ+4wU1lYHLMJmwyQ1GWSKMnXMrVTo+/V9CFeXM/33nvvGYqKJk2aONrAOQSU5iuTJ082JPaZ6kMpojmDMhVJ3fvmm2+cP//5z47O427S6eiF7ZzDTFLKfAc604fzpz/9yQTjMv3SLJImnRHhLPlIFPpoW2cpcXQyyNAuEHtG2qBC3lOocrmRNwKJigXLGyVfReK8ILfXdB6hsWC+KhlP9f/ghsxeL+wGliPzRsuWLfNaA3r//fcdDJBOE+R06tTJueqqqxx+9IUK61baHcHRC+eOHgmYLCHnnntu3moL1Ueg7HHHHRfaPveiXAMKbUhu5IRAIgyQXj9x9FQpBQwLlaSSyVegoLjrrrsi/d9UT6/M/+KMLljEZhTDIvbDDz9sRjEs/PpTCuXafxILnnfeeY7emnb01CyvRfegNumr3qHLOooLqht0LR99YEU2Er0T5uj1IEc7MTokS9S7bI7maDIGOBfDq51S09I0LV++3NHe8kHdlWsFIJAIA6S5jM0owMVJL1Y6Xbt2dU/L5luvUzmk9tFz/tRH+6442gfIYSenUCGTK7nRBg8ebEZwOq10zir5IXqFnTCmN/lKlPo0ZawZlXrxY3SmHQYdovVzEUa42p8qVeV3v/td2n9iqRtyUBAC4gmdxwofntBNmzbN+jnllFNy0k5ALLsreNniF0ROe/x5brzxxowL1JkawScIT+UTTzzRMBeymwY7JAvfOBLaCnpGjhxpnC/xVnaFfuKYyT2yndpK1PpoV695KYJYYcB88803lZ4qK8jv2SHzMmLa9lHKxY9AInbBooaR7Wt21oIEg6EXgBU0q5Szkai3zflxYxiefPJJpcnClF4oN9SzL774otlts+mTv8zo0aOVTu5osmF4d+u0/43Jh4YLAc98zz33+KsGnketj219Qlj88tRTT6VdIjuKSPkgIAYoj3dB8sFBgwZVqcmPnR8WLv/aadD6xx71tjkjNLyq27Rpo4YMGWJcBMh3hrcwfXSF52Cb3kbw/xkzZowaMGBAWnH0kq6H6PWhQ4eaEBCb7emo9eGiQF/Yis+Uwz6t83JScgQSEQ2vd0AUTnO46iM4/uGcyHTEK4cddpjJeeW9ZnNMeh90TZw4UemUzErvNJkwAJu6lGHK4PoqMY055phjzI+6c+fOtirSyvHjRkc2wUjpRe5sxczIRqexNtOZIEc/FBCSoVNfK+21rdq2bZtRJyOlKPXRGDF54K8X8E1iAIwijqE2xjCos9TTqahN5lfukwcNtgRI+L1yySWXeE/lOEcEEmGAMArPPvtsVmj4X1QvNmYt5y2A3tNOO8144PJHmmt2CK8ujokxw2lSc0SbqQ0/JIwJ8WelEr3KaAwGhi3M2U+n6VEEpDIF9U7RgvoctT5vGxjzSZMmGQzJY3b44YcbY47hyNYvrx7Wk7I5IlJe76x5q8lxrggUtISd4MpsybKbpP+oTUYIPXKJHI24ts3z6Sj+PsRbBe3GcU1P9ZxDDjnEWnXU+oIaZodNB6I6OmmkcZzEBUGkvBBIxAgoV6OcrTz/s2rHN+Par53vQkcFUEEQYpGvsK7iLiYzvTjppJNy2rnKt92geh999JHhvyGkgRGldmo06z5gcd9995mpD+tetuEZUesL6jPXtCOmWYxndAqJGLtiImWEQHnZw+rRGz3tMtHpRKhn+uh4opwfSBsdB/2MroiQ1+spxg9Ir1vlrCvqCto9wNlvv/2cBg0apPyUCMfQ0xuHe7lK1Prc9v/5z386I0aMcPS01Tgn6imsg5c5flAi5YWAjIDK4D+DsG1z/IIKGUHF9WgE4TKyIIIfwrR8F3rd/kWhj7UzIthxPdBTL7MQDX4wNerEiW5T8l1mCIgBKoMXArmXu20OhQfTGLaU/ZLLtrm/bk0/xwjC092/f3/jKoDvU5DY0JkE1ZNr8SAgBigeXHPSGvW2eU6N15DCuAfY7FqRgkmkfBAQA1Q+70J6IggkDgGJBUvcK5cHFgTKBwExQOXzLqQngkDiEBADlLhXLg8sCJQPAmKAyuddSE8EgcQhIAYoca9cHlgQKB8ExACVz7uQnggCiUNADFDiXrk8sCBQPgiIASqfdyE9EQQSh4AYoMS9cnlgQaB8EBADVD7vQnoiCCQOATFAiXvl8sCCQPkgIAaofN6F9EQQSBwCYoAS98rlgQWB8kFADFD5vAvpiSCQOAT+H8NjuPaS4KEAAAAAAElFTkSuQmCC" /><!-- --></p>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" tabindex="-1"></a>uFA <span class="ot">&lt;-</span> scores <span class="sc">%*%</span> <span class="fu">t</span>(loadings) <span class="co"># recover UNS effects</span></span>
<span id="cb47-2"><a href="#cb47-2" tabindex="-1"></a>uDG <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(u[,(nPC<span class="sc">+</span><span class="dv">1</span>)<span class="sc">:</span><span class="fu">ncol</span>(u)]) <span class="co"># recover DIAG effects </span></span>
<span id="cb47-3"><a href="#cb47-3" tabindex="-1"></a>u <span class="ot">&lt;-</span> uFA <span class="sc">+</span> uDG <span class="co"># total effects</span></span></code></pre></div>
<p>As can be seen genotype BLUPs for all environments can be recovered
by multiplying the loadings (Gamma) by the factor scores. This is a
parsomonious way to model an unstructured covariance. As an additional
information, notice that we calculate the factor loadings from BLUPs and
the mixed model only calculates the factor scores. This is different to
the asreml software where loadings are calculated as variance components
through REML. Despite the difference we have run multiple examples and
simulations and the BLUPs from both approaches are on average &gt;0.98
correlated so you can be confident that our approach is robust.</p>
</div>
<div id="two-stage-analysis" class="section level2">
<h2>7) Two stage analysis</h2>
<p>It is common then to fit a first model that accounts for the
variation of random design elements, e.g., locations, years, blocks, and
fixed genotype effects to obtain the estimated marginal means (EMMs) or
best linear unbiased estimators (BLUEs) as adjusted entry means. These
adjusted entry means are then used as the phenotype or response variable
in GWAS and genomic prediction studies.</p>
<div class="sourceCode" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" tabindex="-1"></a><span class="do">##########</span></span>
<span id="cb48-2"><a href="#cb48-2" tabindex="-1"></a><span class="do">## stage 1</span></span>
<span id="cb48-3"><a href="#cb48-3" tabindex="-1"></a><span class="do">##########</span></span>
<span id="cb48-4"><a href="#cb48-4" tabindex="-1"></a><span class="fu">data</span>(DT_h2, <span class="at">package=</span><span class="st">&quot;enhancer&quot;</span>)</span>
<span id="cb48-5"><a href="#cb48-5" tabindex="-1"></a>DT <span class="ot">&lt;-</span> DT_h2</span>
<span id="cb48-6"><a href="#cb48-6" tabindex="-1"></a><span class="fu">head</span>(DT)</span></code></pre></div>
<pre><code>##                 Name     Env Loc Year     Block y
## 1            W8822-3 FL.2012  FL 2012 FL.2012.1 2
## 2            W8867-7 FL.2012  FL 2012 FL.2012.2 2
## 3           MSL007-B MO.2011  MO 2011 MO.2011.1 3
## 4         CO00270-7W FL.2012  FL 2012 FL.2012.2 3
## 5 Manistee(MSL292-A) FL.2013  FL 2013 FL.2013.2 3
## 6           MSM246-B FL.2012  FL 2012 FL.2012.2 3</code></pre>
<div class="sourceCode" id="cb50"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" tabindex="-1"></a>envs <span class="ot">&lt;-</span> <span class="fu">unique</span>(DT<span class="sc">$</span>Env)</span>
<span id="cb50-2"><a href="#cb50-2" tabindex="-1"></a>vals <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb50-3"><a href="#cb50-3" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(envs)){</span>
<span id="cb50-4"><a href="#cb50-4" tabindex="-1"></a>  ans1 <span class="ot">&lt;-</span> <span class="fu">lmeb</span>(y<span class="sc">~</span>Name <span class="sc">+</span> (<span class="dv">1</span><span class="sc">|</span>Block), <span class="at">trace =</span> <span class="dv">0</span><span class="dt">L</span>, <span class="at">verbose =</span> <span class="dv">0</span><span class="dt">L</span>,</span>
<span id="cb50-5"><a href="#cb50-5" tabindex="-1"></a>                   <span class="at">data=</span> <span class="fu">droplevels</span>(DT[<span class="fu">which</span>(DT<span class="sc">$</span>Env <span class="sc">==</span> envs[i]),]) )</span>
<span id="cb50-6"><a href="#cb50-6" tabindex="-1"></a>  b <span class="ot">&lt;-</span> <span class="fu">fixef</span>(ans1) </span>
<span id="cb50-7"><a href="#cb50-7" tabindex="-1"></a>  b[<span class="dv">2</span><span class="sc">:</span><span class="fu">length</span>(b)] <span class="ot">&lt;-</span> b[<span class="dv">2</span><span class="sc">:</span><span class="fu">length</span>(b)] <span class="sc">+</span> b[<span class="dv">1</span>]</span>
<span id="cb50-8"><a href="#cb50-8" tabindex="-1"></a>  ids <span class="ot">&lt;-</span> <span class="fu">colnames</span>(<span class="fu">model.matrix</span>(<span class="sc">~</span>Name<span class="dv">-1</span>, <span class="at">data=</span><span class="fu">droplevels</span>(DT[<span class="fu">which</span>(DT<span class="sc">$</span>Env <span class="sc">==</span> envs[i]),]) ))</span>
<span id="cb50-9"><a href="#cb50-9" tabindex="-1"></a>  ids <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">&quot;Name&quot;</span>,<span class="st">&quot;&quot;</span>,ids)</span>
<span id="cb50-10"><a href="#cb50-10" tabindex="-1"></a>  vals[[i]] <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">Estimate=</span>b , <span class="at">stdError=</span> <span class="fu">diag</span>( <span class="fu">vcov</span>(ans1)), <span class="at">Effect=</span> ids, <span class="at">Env=</span> envs[i])</span>
<span id="cb50-11"><a href="#cb50-11" tabindex="-1"></a>}</span></code></pre></div>
<pre><code>## boundary (singular) fit: see help(&#39;isSingular&#39;)
## boundary (singular) fit: see help(&#39;isSingular&#39;)
## boundary (singular) fit: see help(&#39;isSingular&#39;)
## boundary (singular) fit: see help(&#39;isSingular&#39;)
## boundary (singular) fit: see help(&#39;isSingular&#39;)
## boundary (singular) fit: see help(&#39;isSingular&#39;)
## boundary (singular) fit: see help(&#39;isSingular&#39;)</code></pre>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" tabindex="-1"></a>DT2 <span class="ot">&lt;-</span> <span class="fu">do.call</span>(rbind, vals)</span>
<span id="cb52-2"><a href="#cb52-2" tabindex="-1"></a></span>
<span id="cb52-3"><a href="#cb52-3" tabindex="-1"></a><span class="do">##########</span></span>
<span id="cb52-4"><a href="#cb52-4" tabindex="-1"></a><span class="do">## stage 2</span></span>
<span id="cb52-5"><a href="#cb52-5" tabindex="-1"></a><span class="do">##########</span></span>
<span id="cb52-6"><a href="#cb52-6" tabindex="-1"></a>DT2<span class="sc">$</span>w <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">/</span>DT2<span class="sc">$</span>stdError</span>
<span id="cb52-7"><a href="#cb52-7" tabindex="-1"></a>ans2 <span class="ot">&lt;-</span> <span class="fu">lmeb</span>(Estimate<span class="sc">~</span>Env <span class="sc">+</span> (<span class="dv">1</span><span class="sc">|</span>Effect) <span class="sc">+</span> (<span class="dv">1</span><span class="sc">|</span>Env<span class="sc">:</span>Effect), <span class="at">weights =</span> w,</span>
<span id="cb52-8"><a href="#cb52-8" tabindex="-1"></a>                 <span class="at">data=</span>DT2,  <span class="at">verbose =</span> <span class="dv">0</span><span class="dt">L</span>, <span class="at">trace=</span><span class="dv">0</span><span class="dt">L</span></span>
<span id="cb52-9"><a href="#cb52-9" tabindex="-1"></a>                 )</span>
<span id="cb52-10"><a href="#cb52-10" tabindex="-1"></a>vc <span class="ot">&lt;-</span> <span class="fu">VarCorr</span>(ans2); <span class="fu">print</span>(vc,<span class="at">comp=</span><span class="fu">c</span>(<span class="st">&quot;Variance&quot;</span>))</span></code></pre></div>
<pre><code>##  Groups     Name        Variance
##  Env:Effect (Intercept) 2.91649 
##  Effect     (Intercept) 1.99225 
##  Residual               0.72139</code></pre>
<div class="sourceCode" id="cb54"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" tabindex="-1"></a>ve <span class="ot">&lt;-</span> <span class="fu">attr</span>(vc, <span class="st">&quot;sc&quot;</span>)<span class="sc">^</span><span class="dv">2</span>; ve</span></code></pre></div>
<pre><code>## [1] 0.7213944</code></pre>
</div>
<div id="how-to-deal-with-massive-gxe-models-or-a-massive-number-of-individuals" class="section level2">
<h2>8) How to deal with massive GxE models or a massive number of
individuals</h2>
<p>Up to now, the problem of dealing with a massive number of
environments and genotyped individuals is a headeache for everyone. In
this section i just aim to give you some ideas of what you could do to
fit massive models with limited trade-off in accuracy.</p>
<p>The lme4breeding package includes the argument ‘rotation’ to
implement the proposed eigen decomposition of the genomic relationship
matrix and be able to fit a fully sparse mixed model (Lee and Van der
Werf, 2006). This trick can be used in any software that can take the
relationship matrix. The main inconvenience is that the current
mathematical derivation only allows for a single random effect and this
needs to be balanced and complete (no missing data). Still extremely
powerful:</p>
<div class="sourceCode" id="cb56"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" tabindex="-1"></a><span class="co"># load data (phenotypes and genotypes)</span></span>
<span id="cb56-2"><a href="#cb56-2" tabindex="-1"></a><span class="fu">data</span>(DT_big, <span class="at">package=</span><span class="st">&quot;enhancer&quot;</span>)</span>
<span id="cb56-3"><a href="#cb56-3" tabindex="-1"></a>DT <span class="ot">=</span> DT_big</span>
<span id="cb56-4"><a href="#cb56-4" tabindex="-1"></a>M <span class="ot">=</span> <span class="fu">apply</span>(M_big,<span class="dv">2</span>,as.numeric) <span class="co"># change from bits to numeric</span></span>
<span id="cb56-5"><a href="#cb56-5" tabindex="-1"></a><span class="fu">rownames</span>(M) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(M_big)</span>
<span id="cb56-6"><a href="#cb56-6" tabindex="-1"></a>DT[,<span class="st">&quot;envf_repf&quot;</span>] <span class="ot">=</span> <span class="fu">paste</span>(DT[,<span class="st">&quot;envf&quot;</span>],DT[,<span class="st">&quot;repf&quot;</span>],<span class="at">sep=</span><span class="st">&quot;_&quot;</span>)</span>
<span id="cb56-7"><a href="#cb56-7" tabindex="-1"></a>DT <span class="ot">&lt;-</span> DT[<span class="fu">with</span>(DT, <span class="fu">order</span>(envf_repf, id)), ]</span>
<span id="cb56-8"><a href="#cb56-8" tabindex="-1"></a></span>
<span id="cb56-9"><a href="#cb56-9" tabindex="-1"></a><span class="co"># we will make up missing data</span></span>
<span id="cb56-10"><a href="#cb56-10" tabindex="-1"></a>indsNa <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="fu">unique</span>(DT<span class="sc">$</span>id), <span class="dv">200</span>)</span>
<span id="cb56-11"><a href="#cb56-11" tabindex="-1"></a>nas <span class="ot">&lt;-</span> <span class="fu">which</span>(DT<span class="sc">$</span>id <span class="sc">%in%</span> indsNa) <span class="co"># sample(1:nrow(DT), round(nrow(DT)*.25))</span></span>
<span id="cb56-12"><a href="#cb56-12" tabindex="-1"></a>DT<span class="sc">$</span>value[nas]<span class="ot">=</span><span class="cn">NA</span></span>
<span id="cb56-13"><a href="#cb56-13" tabindex="-1"></a>DT<span class="sc">$</span>value <span class="ot">&lt;-</span> <span class="fu">imputev</span>(DT<span class="sc">$</span>value, <span class="at">by=</span>DT<span class="sc">$</span>envf_repf)</span>
<span id="cb56-14"><a href="#cb56-14" tabindex="-1"></a>DT<span class="sc">$</span>nas <span class="ot">&lt;-</span> <span class="dv">0</span>; DT<span class="sc">$</span>nas[nas]<span class="ot">=</span><span class="dv">1</span></span>
<span id="cb56-15"><a href="#cb56-15" tabindex="-1"></a></span>
<span id="cb56-16"><a href="#cb56-16" tabindex="-1"></a><span class="co"># compute the relationship matrix</span></span>
<span id="cb56-17"><a href="#cb56-17" tabindex="-1"></a>MMT <span class="ot">&lt;-</span> <span class="fu">tcrossprod</span>(M)</span>
<span id="cb56-18"><a href="#cb56-18" tabindex="-1"></a>m <span class="ot">&lt;-</span> <span class="fu">sum</span>(<span class="fu">diag</span>(MMT))<span class="sc">/</span><span class="fu">nrow</span>(MMT)</span>
<span id="cb56-19"><a href="#cb56-19" tabindex="-1"></a>MMT <span class="ot">&lt;-</span> MMT<span class="sc">/</span>m</span>
<span id="cb56-20"><a href="#cb56-20" tabindex="-1"></a>MMT <span class="ot">&lt;-</span> MMT <span class="sc">+</span> <span class="fu">diag</span>(<span class="fl">1e-05</span>, <span class="fu">ncol</span>(MMT), <span class="fu">ncol</span>(MMT))</span>
<span id="cb56-21"><a href="#cb56-21" tabindex="-1"></a></span>
<span id="cb56-22"><a href="#cb56-22" tabindex="-1"></a><span class="do">## Fit the main effect model</span></span>
<span id="cb56-23"><a href="#cb56-23" tabindex="-1"></a>lmod <span class="ot">&lt;-</span> <span class="fu">lmeb</span>(<span class="at">formula=</span>value<span class="sc">~</span>(<span class="dv">1</span><span class="sc">|</span>id), <span class="at">data =</span> DT, </span>
<span id="cb56-24"><a href="#cb56-24" tabindex="-1"></a>                  <span class="at">relmat =</span> <span class="fu">list</span>(<span class="at">id =</span> MMT), </span>
<span id="cb56-25"><a href="#cb56-25" tabindex="-1"></a>                  <span class="at">verbose =</span> <span class="dv">0</span><span class="dt">L</span>, <span class="at">trace=</span><span class="dv">0</span><span class="dt">L</span>, <span class="at">rotation=</span><span class="cn">TRUE</span></span>
<span id="cb56-26"><a href="#cb56-26" tabindex="-1"></a>)</span>
<span id="cb56-27"><a href="#cb56-27" tabindex="-1"></a></span>
<span id="cb56-28"><a href="#cb56-28" tabindex="-1"></a><span class="co"># extract GEBVs</span></span>
<span id="cb56-29"><a href="#cb56-29" tabindex="-1"></a>ran0 <span class="ot">&lt;-</span> <span class="fu">ranef</span>(lmod, <span class="at">condVar=</span><span class="cn">FALSE</span>)</span></code></pre></div>
<pre><code>## Rotating back BLUPs by transpose of Cholesky (u=L&#39;u*) for: id</code></pre>
<div class="sourceCode" id="cb58"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" tabindex="-1"></a>xx<span class="ot">=</span><span class="fu">as.data.frame</span>(<span class="fu">as.matrix</span>(ran0<span class="sc">$</span>id) <span class="sc">%*%</span> <span class="fu">matrix</span>(<span class="fu">rep</span>(<span class="dv">1</span>,<span class="dv">50</span>), <span class="at">nrow=</span><span class="dv">1</span>) )</span>
<span id="cb58-2"><a href="#cb58-2" tabindex="-1"></a>xx<span class="sc">$</span>id <span class="ot">&lt;-</span> <span class="fu">rownames</span>(xx)</span>
<span id="cb58-3"><a href="#cb58-3" tabindex="-1"></a></span>
<span id="cb58-4"><a href="#cb58-4" tabindex="-1"></a><span class="co"># Compare against</span></span>
<span id="cb58-5"><a href="#cb58-5" tabindex="-1"></a>xxl <span class="ot">&lt;-</span> <span class="fu">reshape</span>(xx, <span class="at">idvar =</span> <span class="st">&quot;id&quot;</span>, <span class="at">varying =</span> <span class="fu">list</span>(<span class="dv">1</span><span class="sc">:</span>(<span class="fu">ncol</span>(xx)<span class="sc">-</span><span class="dv">1</span>)),</span>
<span id="cb58-6"><a href="#cb58-6" tabindex="-1"></a>               <span class="at">v.names =</span> <span class="st">&quot;gebv&quot;</span>, <span class="at">direction =</span> <span class="st">&quot;long&quot;</span>, <span class="at">timevar =</span> <span class="st">&quot;envf_repf&quot;</span>,</span>
<span id="cb58-7"><a href="#cb58-7" tabindex="-1"></a>               <span class="at">times =</span> <span class="fu">unique</span>(DT<span class="sc">$</span>envf_repf) )</span>
<span id="cb58-8"><a href="#cb58-8" tabindex="-1"></a>DTc <span class="ot">&lt;-</span> <span class="fu">merge</span>(DT, xxl, <span class="at">by=</span><span class="fu">c</span>(<span class="st">&quot;id&quot;</span>,<span class="st">&quot;envf_repf&quot;</span>), <span class="at">all.x =</span> <span class="cn">TRUE</span>)</span>
<span id="cb58-9"><a href="#cb58-9" tabindex="-1"></a><span class="co"># with(DTc[which(DTc$nas ==1),], plot(gebv,gv))</span></span>
<span id="cb58-10"><a href="#cb58-10" tabindex="-1"></a><span class="fu">with</span>(DTc[<span class="fu">which</span>(DTc<span class="sc">$</span>nas <span class="sc">==</span><span class="dv">1</span>),], <span class="fu">cor</span>(gebv,gv)) <span class="co"># accuracy is 0.39 </span></span></code></pre></div>
<pre><code>## [1] 0.5216427</code></pre>
<p>Next, we have to deal with a big number of environments (in the order
of hundreds). This requires a massive amount of computational time when
the genomic relationship matrix needs to be included. My proposal
(please cite this vignette if you use it since I am not planning to
publish it), is:</p>
<ol style="list-style-type: decimal">
<li>Define the unit of environment to balance (e.g., environment by
replicate)</li>
<li>Make your dataset fully balanced for the environment unit using the
‘balanceData’ function (each individual should show up everywhere)</li>
<li>Impute with the mean value of each environment the missing-data
cells (e.g., using the imputev function)</li>
<li>Perform the eigen decomposition of the relationship matrix as define
at the beginning of this section using the rotation argument.</li>
<li>Run a GxE diagonal model with the argument rotation=TRUE</li>
<li>Use the results to run more complex models like heterogeneous
residual variance or a reduced rank GxE model (e.g., 3 factors) as
specified in the section 6 of this vignette.</li>
</ol>
<p>My simulations show that this model is extremely accurate to predict
the environment-specific breeding values along with the
across-environment breeding values. The computation time is extremely
low and the accuracy is comparable to the classical model which would
take hundreds of times more computation to get similar results</p>
<div class="sourceCode" id="cb60"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" tabindex="-1"></a><span class="do">## Fit the diagonal model</span></span>
<span id="cb60-2"><a href="#cb60-2" tabindex="-1"></a>lmod <span class="ot">&lt;-</span> <span class="fu">lmeb</span>(<span class="at">formula=</span>value<span class="sc">~</span>(<span class="dv">0</span><span class="sc">+</span>envf_repf<span class="sc">||</span>id), <span class="at">data =</span> DT, </span>
<span id="cb60-3"><a href="#cb60-3" tabindex="-1"></a>                 <span class="at">relmat =</span> <span class="fu">list</span>(<span class="at">id =</span> MMT), <span class="at">nIters =</span> <span class="dv">1000</span>,</span>
<span id="cb60-4"><a href="#cb60-4" tabindex="-1"></a>                 <span class="at">verbose =</span> <span class="dv">0</span><span class="dt">L</span>, <span class="at">trace=</span><span class="dv">0</span><span class="dt">L</span>, <span class="at">rotation=</span><span class="cn">TRUE</span></span>
<span id="cb60-5"><a href="#cb60-5" tabindex="-1"></a>)</span>
<span id="cb60-6"><a href="#cb60-6" tabindex="-1"></a></span>
<span id="cb60-7"><a href="#cb60-7" tabindex="-1"></a><span class="co"># Extract GEBVs</span></span>
<span id="cb60-8"><a href="#cb60-8" tabindex="-1"></a>ran0 <span class="ot">&lt;-</span> <span class="fu">ranef</span>(lmod, <span class="at">condVar=</span><span class="cn">FALSE</span>)</span></code></pre></div>
<pre><code>## Rotating back BLUPs by transpose of Cholesky (u=L&#39;u*) for: id</code></pre>
<div class="sourceCode" id="cb62"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" tabindex="-1"></a>xx<span class="ot">=</span><span class="fu">as.data.frame</span>(ran0<span class="sc">$</span>id)<span class="co"># u=H0</span></span>
<span id="cb62-2"><a href="#cb62-2" tabindex="-1"></a>xx<span class="sc">$</span>id <span class="ot">&lt;-</span> <span class="fu">rownames</span>(xx)</span>
<span id="cb62-3"><a href="#cb62-3" tabindex="-1"></a></span>
<span id="cb62-4"><a href="#cb62-4" tabindex="-1"></a><span class="co"># Compare against</span></span>
<span id="cb62-5"><a href="#cb62-5" tabindex="-1"></a>xxl <span class="ot">&lt;-</span> <span class="fu">reshape</span>(xx, <span class="at">idvar =</span> <span class="st">&quot;id&quot;</span>, <span class="at">varying =</span> <span class="fu">list</span>(<span class="dv">1</span><span class="sc">:</span>(<span class="fu">ncol</span>(xx)<span class="sc">-</span><span class="dv">1</span>)),</span>
<span id="cb62-6"><a href="#cb62-6" tabindex="-1"></a>               <span class="at">v.names =</span> <span class="st">&quot;gebv&quot;</span>, <span class="at">direction =</span> <span class="st">&quot;long&quot;</span>, <span class="at">timevar =</span> <span class="st">&quot;envf_repf&quot;</span>,</span>
<span id="cb62-7"><a href="#cb62-7" tabindex="-1"></a>               <span class="at">times =</span> <span class="fu">colnames</span>(xx)[<span class="dv">1</span><span class="sc">:</span>(<span class="fu">ncol</span>(xx)<span class="sc">-</span><span class="dv">1</span>)])</span>
<span id="cb62-8"><a href="#cb62-8" tabindex="-1"></a>DTc <span class="ot">&lt;-</span> <span class="fu">merge</span>(DT, xxl, <span class="at">by=</span><span class="fu">c</span>(<span class="st">&quot;id&quot;</span>,<span class="st">&quot;envf_repf&quot;</span>), <span class="at">all.x =</span> <span class="cn">TRUE</span>)</span>
<span id="cb62-9"><a href="#cb62-9" tabindex="-1"></a><span class="co"># with(DTc[which(DTc$nas ==1),], plot(gebv,gv))</span></span>
<span id="cb62-10"><a href="#cb62-10" tabindex="-1"></a><span class="fu">with</span>(DTc[<span class="fu">which</span>(DTc<span class="sc">$</span>nas <span class="sc">==</span><span class="dv">1</span>),], <span class="fu">cor</span>(gebv,gv)) <span class="co"># accuracy is ~0.62</span></span></code></pre></div>
<pre><code>## [1] 0.6846091</code></pre>
<p>Additionally, you could fit a different residual variance at each
environments by creating the record number variable as a factor. Among
the several ways you could do that here is one:</p>
<div class="sourceCode" id="cb64"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" tabindex="-1"></a><span class="co"># lmodRes &lt;- lmeb(formula=value~(0+envf_repf||id) + (0+envf_repf||unitsR), data = DT,</span></span>
<span id="cb64-2"><a href="#cb64-2" tabindex="-1"></a><span class="co">#                  relmat = list(id = MMT), nIters=1000,</span></span>
<span id="cb64-3"><a href="#cb64-3" tabindex="-1"></a><span class="co">#                  verbose = 0L, trace=0L, rotation=TRUE</span></span>
<span id="cb64-4"><a href="#cb64-4" tabindex="-1"></a><span class="co"># )</span></span>
<span id="cb64-5"><a href="#cb64-5" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb64-6"><a href="#cb64-6" tabindex="-1"></a><span class="co"># # Extract GEBVs</span></span>
<span id="cb64-7"><a href="#cb64-7" tabindex="-1"></a><span class="co"># ran0 &lt;- ranef(lmodRes, condVar=FALSE)</span></span>
<span id="cb64-8"><a href="#cb64-8" tabindex="-1"></a><span class="co"># xx=as.data.frame(ran0$id)# u=H0</span></span>
<span id="cb64-9"><a href="#cb64-9" tabindex="-1"></a><span class="co"># xx$id &lt;- rownames(xx)</span></span>
<span id="cb64-10"><a href="#cb64-10" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb64-11"><a href="#cb64-11" tabindex="-1"></a><span class="co"># # Compare against</span></span>
<span id="cb64-12"><a href="#cb64-12" tabindex="-1"></a><span class="co"># xxl &lt;- reshape(xx, idvar = &quot;id&quot;, varying = list(1:(ncol(xx)-1)),</span></span>
<span id="cb64-13"><a href="#cb64-13" tabindex="-1"></a><span class="co">#                v.names = &quot;gebv&quot;, direction = &quot;long&quot;, timevar = &quot;envf_repf&quot;,</span></span>
<span id="cb64-14"><a href="#cb64-14" tabindex="-1"></a><span class="co">#                times = colnames(xx)[1:(ncol(xx)-1)])</span></span>
<span id="cb64-15"><a href="#cb64-15" tabindex="-1"></a><span class="co"># DTc &lt;- merge(DT, xxl, by=c(&quot;id&quot;,&quot;envf_repf&quot;), all.x = TRUE)</span></span>
<span id="cb64-16"><a href="#cb64-16" tabindex="-1"></a><span class="co"># with(DTc[which(DTc$nas ==1),], plot(gebv,gv))</span></span>
<span id="cb64-17"><a href="#cb64-17" tabindex="-1"></a><span class="co"># with(DTc[which(DTc$nas ==1),], cor(gebv,gv)) # accuracy is ~0.62</span></span></code></pre></div>
<p>Last but not least, you can use the results from the diagonal model
to fit a reduced rank model:</p>
<div class="sourceCode" id="cb65"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" tabindex="-1"></a><span class="co"># # get latent covariates and add them to the dataset</span></span>
<span id="cb65-2"><a href="#cb65-2" tabindex="-1"></a><span class="co"># Gammas &lt;- with(DT,  rrm(envf_repf, H = ran0$id, nPC = 3, returnGamma = TRUE))</span></span>
<span id="cb65-3"><a href="#cb65-3" tabindex="-1"></a><span class="co"># Zrr &lt;- Gammas$Zstar</span></span>
<span id="cb65-4"><a href="#cb65-4" tabindex="-1"></a><span class="co"># for(i in 1:ncol(Zrr)){DT[,colnames(Zrr)[i]] &lt;- Zrr[,i]}</span></span>
<span id="cb65-5"><a href="#cb65-5" tabindex="-1"></a><span class="co"># # fit the model</span></span>
<span id="cb65-6"><a href="#cb65-6" tabindex="-1"></a><span class="co"># ansRR &lt;- lmeb(value~(0+PC1+PC2+PC3|id) + (0+envf_repf||id), data = DT, </span></span>
<span id="cb65-7"><a href="#cb65-7" tabindex="-1"></a><span class="co">#                  relmat = list(id = MMT), nIters = 1000,</span></span>
<span id="cb65-8"><a href="#cb65-8" tabindex="-1"></a><span class="co">#                  verbose = 0L, trace=0L, rotation=TRUE</span></span>
<span id="cb65-9"><a href="#cb65-9" tabindex="-1"></a><span class="co"># )</span></span>
<span id="cb65-10"><a href="#cb65-10" tabindex="-1"></a><span class="co"># vc &lt;- VarCorr(ansRR); print(vc,comp=c(&quot;Variance&quot;))</span></span>
<span id="cb65-11"><a href="#cb65-11" tabindex="-1"></a><span class="co"># ve &lt;- attr(vc, &quot;sc&quot;)^2; ve</span></span>
<span id="cb65-12"><a href="#cb65-12" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb65-13"><a href="#cb65-13" tabindex="-1"></a><span class="co"># loadings=Gammas$Gamma</span></span>
<span id="cb65-14"><a href="#cb65-14" tabindex="-1"></a><span class="co"># Gint &lt;- loadings %*% vc$id %*% t(loadings)</span></span>
<span id="cb65-15"><a href="#cb65-15" tabindex="-1"></a><span class="co"># Gspec &lt;- diag( unlist(lapply(vc[2:(length(vc))], function(x){x[[1]]})) )</span></span>
<span id="cb65-16"><a href="#cb65-16" tabindex="-1"></a><span class="co"># G &lt;- Gint + Gspec</span></span>
<span id="cb65-17"><a href="#cb65-17" tabindex="-1"></a><span class="co"># u &lt;- ranef(ansRR, condVar=FALSE)$id</span></span>
<span id="cb65-18"><a href="#cb65-18" tabindex="-1"></a><span class="co"># uInter &lt;- as.matrix(u[,1:3]) %*% t(as.matrix(loadings))</span></span>
<span id="cb65-19"><a href="#cb65-19" tabindex="-1"></a><span class="co"># uSpec &lt;- as.matrix(u[,-c(1:3)])</span></span>
<span id="cb65-20"><a href="#cb65-20" tabindex="-1"></a><span class="co"># u &lt;- uSpec + uInter</span></span>
<span id="cb65-21"><a href="#cb65-21" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb65-22"><a href="#cb65-22" tabindex="-1"></a><span class="co"># xx=as.data.frame(u)# u=H0</span></span>
<span id="cb65-23"><a href="#cb65-23" tabindex="-1"></a><span class="co"># xx$id &lt;- rownames(xx)</span></span>
<span id="cb65-24"><a href="#cb65-24" tabindex="-1"></a><span class="co"># xxl &lt;- reshape(xx, idvar = &quot;id&quot;, varying = list(1:(ncol(xx)-1)),</span></span>
<span id="cb65-25"><a href="#cb65-25" tabindex="-1"></a><span class="co">#                v.names = &quot;gebv&quot;, direction = &quot;long&quot;, timevar = &quot;envf_repf&quot;,</span></span>
<span id="cb65-26"><a href="#cb65-26" tabindex="-1"></a><span class="co">#                times = colnames(xx)[1:(ncol(xx)-1)])</span></span>
<span id="cb65-27"><a href="#cb65-27" tabindex="-1"></a><span class="co"># DTc &lt;- merge(DT, xxl, by=c(&quot;id&quot;,&quot;envf_repf&quot;), all.x = TRUE)</span></span>
<span id="cb65-28"><a href="#cb65-28" tabindex="-1"></a><span class="co"># with(DTc[which(DTc$nas ==1),], plot(gebv,gv))</span></span>
<span id="cb65-29"><a href="#cb65-29" tabindex="-1"></a><span class="co"># with(DTc[which(DTc$nas ==1),], cor(gebv,gv)) # accuracy is 0.62</span></span></code></pre></div>
<p>For the sake of time we are not running the last two models. Also
because the heterogeneous residual model and reduced rank models do not
increase the accuracy in this particular example since the simulated
data didn’t take additional noise factors into account.</p>
<p>Notice that we have fitted a multi-environmental model with 1K
individuals in 25 environments, with 2 replicates in each environment
with a genomic relationship matrix in a few seconds. It hardly gets any
better than this.</p>
</div>
<div id="literature" class="section level2">
<h2>Literature</h2>
<p>Giovanny Covarrubias-Pazaran (2024). lme4breeding: enabling genetic
evaluation in the age of genomic data. To be submitted.</p>
<p>Bates Douglas, Maechler Martin, Bolker Ben, Walker Steve. 2015.
Fitting Linear Mixed-Effects Models Using lme4. Journal of Statistical
Software, 67(1), 1-48.</p>
<p>Bernardo Rex. 2010. Breeding for quantitative traits in plants.
Second edition. Stemma Press. 390 pp.</p>
<p>Henderson C.R. 1975. Best Linear Unbiased Estimation and Prediction
under a Selection Model. Biometrics vol. 31(2):423-447.</p>
<p>Searle. 1993. Applying the EM algorithm to calculating ML and REML
estimates of variance components. Paper invited for the 1993 American
Statistical Association Meeting, San Francisco.</p>
<p>Welham, S., Cullis, B., Gogel, B., Gilmour, A., and Thompson, R.
(2004). Prediction in linear mixed models. Australian and New Zealand
Journal of Statistics, 46, 325 - 347.</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
